// rolldown@1.0.0-beta.8-commit.852c603/dist/shared/filter-expression-index-CRtoeipP.cjs
"use strict";

//#region src/filter-expression-index.ts
var And = class {
	kind;
	left;
	right;
	constructor(left, right) {
		this.left = left;
		this.right = right;
		this.kind = "and";
	}
};
function and(left, right) {
	return {
		kind: "and",
		left,
		right
	};
}
function or(left, right) {
	return {
		kind: "or",
		left,
		right
	};
}
function not(expr) {
	return {
		kind: "not",
		expr
	};
}
function id(pattern) {
	return {
		kind: "id",
		pattern
	};
}
function moduleType(pattern) {
	return {
		kind: "moduleType",
		pattern
	};
}
function code(pattern) {
	return {
		kind: "code",
		pattern
	};
}
function include(expr) {
	return {
		kind: "include",
		expr
	};
}
function exclude(expr) {
	return {
		kind: "exclude",
		expr
	};
}

//#endregion
Object.defineProperty(exports, 'And', {
  enumerable: true,
  get: function () {
    return And;
  }
});
Object.defineProperty(exports, 'and', {
  enumerable: true,
  get: function () {
    return and;
  }
});
Object.defineProperty(exports, 'code', {
  enumerable: true,
  get: function () {
    return code;
  }
});
Object.defineProperty(exports, 'exclude', {
  enumerable: true,
  get: function () {
    return exclude;
  }
});
Object.defineProperty(exports, 'id', {
  enumerable: true,
  get: function () {
    return id;
  }
});
Object.defineProperty(exports, 'include', {
  enumerable: true,
  get: function () {
    return include;
  }
});
Object.defineProperty(exports, 'moduleType', {
  enumerable: true,
  get: function () {
    return moduleType;
  }
});
Object.defineProperty(exports, 'not', {
  enumerable: true,
  get: function () {
    return not;
  }
});
Object.defineProperty(exports, 'or', {
  enumerable: true,
  get: function () {
    return or;
  }
});
