// rolldown@1.0.0-beta.8-commit.852c603/dist/shared/src-CaxK4-gB.mjs
import { __esm } from "./chunk-DUYDk_2O.mjs";
import { augmentCodeLocation, error, import_binding, init_logs, init_parse_ast_index, logCycleLoading, logInputHookInOutputPlugin, logInvalidLogPosition, logMultiplyNotifyOption, logPluginError, parseAst } from "./parse-ast-index-D0SmlXT6.mjs";
import { and, code, exclude, id, include, init_filter_expression_index, moduleType, or } from "./filter-expression-index-BKmgnKlV.mjs";
import path from "node:path";
import { toJsonSchema } from "@valibot/to-json-schema";
import colors from "ansis";
import * as v from "valibot";
import { availableParallelism } from "node:os";
import { Worker } from "node:worker_threads";

//#region package.json
var version = "1.0.0-beta.8-commit.852c603";
var description = "Fast JavaScript/TypeScript bundler in Rust with Rollup-compatible API.";

//#endregion
//#region src/builtin-plugin/utils.ts
function makeBuiltinPluginCallable(plugin) {
	let callablePlugin = new import_binding.BindingCallableBuiltinPlugin(bindingifyBuiltInPlugin(plugin));
	const wrappedPlugin = plugin;
	for (const key in callablePlugin) wrappedPlugin[key] = function(...args) {
		return callablePlugin[key](...args);
	};
	return wrappedPlugin;
}
function bindingifyBuiltInPlugin(plugin) {
	return {
		__name: plugin.name,
		options: plugin._options
	};
}
var init_utils = __esm({ "src/builtin-plugin/utils.ts"() {} });

//#endregion
//#region src/builtin-plugin/constructors.ts
function modulePreloadPolyfillPlugin(config) {
	return new BuiltinPlugin("builtin:module-preload-polyfill", config);
}
function dynamicImportVarsPlugin(config) {
	return new BuiltinPlugin("builtin:dynamic-import-vars", config);
}
function importGlobPlugin(config) {
	return new BuiltinPlugin("builtin:import-glob", config);
}
function reportPlugin(config) {
	return new BuiltinPlugin("builtin:report", config);
}
function manifestPlugin(config) {
	return new BuiltinPlugin("builtin:manifest", config);
}
function wasmHelperPlugin() {
	return new BuiltinPlugin("builtin:wasm-helper");
}
function wasmFallbackPlugin() {
	return new BuiltinPlugin("builtin:wasm-fallback");
}
function loadFallbackPlugin() {
	return new BuiltinPlugin("builtin:load-fallback");
}
function jsonPlugin(config) {
	return new BuiltinPlugin("builtin:json", config);
}
function buildImportAnalysisPlugin(config) {
	return new BuiltinPlugin("builtin:build-import-analysis", config);
}
function viteResolvePlugin(config) {
	const builtinPlugin = new BuiltinPlugin("builtin:vite-resolve", config);
	return makeBuiltinPluginCallable(builtinPlugin);
}
function moduleFederationPlugin(config) {
	return new BuiltinPlugin("builtin:module-federation", {
		...config,
		remotes: config.remotes && Object.entries(config.remotes).map(([name, remote]) => {
			if (typeof remote === "string") {
				const [entryGlobalName] = remote.split("@");
				const entry = remote.replace(entryGlobalName + "@", "");
				return {
					entry,
					name,
					entryGlobalName
				};
			}
			return {
				...remote,
				name: remote.name ?? name
			};
		}),
		manifest: config.manifest === false ? void 0 : config.manifest === true ? {} : config.manifest
	});
}
function isolatedDeclarationPlugin(config) {
	return new BuiltinPlugin("builtin:isolated-declaration", config);
}
var BuiltinPlugin;
var init_constructors = __esm({ "src/builtin-plugin/constructors.ts"() {
	init_utils();
	BuiltinPlugin = class {
		constructor(name, _options) {
			this.name = name;
			this._options = _options;
		}
	};
} });

//#endregion
//#region src/utils/misc.ts
function arraify(value) {
	return Array.isArray(value) ? value : [value];
}
function isNullish(value) {
	return value === null || value === void 0;
}
function isPromiseLike(value) {
	return value && (typeof value === "object" || typeof value === "function") && typeof value.then === "function";
}
function unimplemented(info) {
	if (info) throw new Error(`unimplemented: ${info}`);
	throw new Error("unimplemented");
}
function unreachable(info) {
	if (info) throw new Error(`unreachable: ${info}`);
	throw new Error("unreachable");
}
function unsupported(info) {
	throw new Error(`UNSUPPORTED: ${info}`);
}
function noop(..._args) {}
var init_misc = __esm({ "src/utils/misc.ts"() {} });

//#endregion
//#region src/log/logging.ts
var LOG_LEVEL_SILENT, LOG_LEVEL_ERROR, LOG_LEVEL_WARN, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, logLevelPriority;
var init_logging = __esm({ "src/log/logging.ts"() {
	LOG_LEVEL_SILENT = "silent";
	LOG_LEVEL_ERROR = "error";
	LOG_LEVEL_WARN = "warn";
	LOG_LEVEL_INFO = "info";
	LOG_LEVEL_DEBUG = "debug";
	logLevelPriority = {
		[LOG_LEVEL_DEBUG]: 0,
		[LOG_LEVEL_INFO]: 1,
		[LOG_LEVEL_WARN]: 2,
		[LOG_LEVEL_SILENT]: 3
	};
} });

//#endregion
//#region src/log/log-handler.ts
function getLogHandler(level, code$1, logger, pluginName, logLevel) {
	if (logLevelPriority[level] < logLevelPriority[logLevel]) return noop;
	return (log, pos) => {
		if (pos != null) logger(LOG_LEVEL_WARN, logInvalidLogPosition(pluginName));
		log = normalizeLog(log);
		if (log.code && !log.pluginCode) log.pluginCode = log.code;
		log.code = code$1;
		log.plugin = pluginName;
		logger(level, log);
	};
}
var normalizeLog;
var init_log_handler = __esm({ "src/log/log-handler.ts"() {
	init_misc();
	init_logging();
	init_logs();
	normalizeLog = (log) => typeof log === "string" ? { message: log } : typeof log === "function" ? normalizeLog(log()) : log;
} });

//#endregion
//#region src/log/logger.ts
function getLogger(plugins, onLog, logLevel, watchMode) {
	const minimalPriority = logLevelPriority[logLevel];
	const logger = (level, log, skipped = new Set()) => {
		const logPriority = logLevelPriority[level];
		if (logPriority < minimalPriority) return;
		for (const plugin of getSortedPlugins("onLog", plugins)) {
			if (skipped.has(plugin)) continue;
			const { onLog: pluginOnLog } = plugin;
			if (pluginOnLog) {
				const getLogHandler$1 = (level$1) => {
					if (logLevelPriority[level$1] < minimalPriority) return () => {};
					return (log$1) => logger(level$1, normalizeLog(log$1), new Set(skipped).add(plugin));
				};
				const handler = "handler" in pluginOnLog ? pluginOnLog.handler : pluginOnLog;
				if (handler.call({
					debug: getLogHandler$1(LOG_LEVEL_DEBUG),
					error: (log$1) => error(normalizeLog(log$1)),
					info: getLogHandler$1(LOG_LEVEL_INFO),
					meta: {
						rollupVersion: "4.23.0",
						rolldownVersion: VERSION,
						watchMode
					},
					warn: getLogHandler$1(LOG_LEVEL_WARN),
					pluginName: plugin.name || "unknown"
				}, level, log) === false) return;
			}
		}
		onLog(level, log);
	};
	return logger;
}
function relativeId(id$1) {
	if (!path.isAbsolute(id$1)) return id$1;
	return path.relative(path.resolve(), id$1);
}
var getOnLog, getDefaultOnLog, addLogToString, defaultPrintLog, getExtendedLogMessage;
var init_logger = __esm({ "src/log/logger.ts"() {
	init_src();
	init_plugin_driver();
	init_log_handler();
	init_logging();
	init_logs();
	getOnLog = (config, logLevel, printLog = defaultPrintLog) => {
		const { onwarn, onLog } = config;
		const defaultOnLog = getDefaultOnLog(printLog, onwarn);
		if (onLog) {
			const minimalPriority = logLevelPriority[logLevel];
			return (level, log) => onLog(level, addLogToString(log), (level$1, handledLog) => {
				if (level$1 === LOG_LEVEL_ERROR) return error(normalizeLog(handledLog));
				if (logLevelPriority[level$1] >= minimalPriority) defaultOnLog(level$1, normalizeLog(handledLog));
			});
		}
		return defaultOnLog;
	};
	getDefaultOnLog = (printLog, onwarn) => onwarn ? (level, log) => {
		if (level === LOG_LEVEL_WARN) onwarn(addLogToString(log), (warning) => printLog(LOG_LEVEL_WARN, normalizeLog(warning)));
		else printLog(level, log);
	} : printLog;
	addLogToString = (log) => {
		Object.defineProperty(log, "toString", {
			value: () => getExtendedLogMessage(log),
			writable: true
		});
		return log;
	};
	defaultPrintLog = (level, log) => {
		const message = getExtendedLogMessage(log);
		switch (level) {
			case LOG_LEVEL_WARN: return console.warn(message);
			case LOG_LEVEL_DEBUG: return console.debug(message);
			default: return console.info(message);
		}
	};
	getExtendedLogMessage = (log) => {
		let prefix = "";
		if (log.plugin) prefix += `(${log.plugin} plugin) `;
		if (log.loc) prefix += `${relativeId(log.loc.file)} (${log.loc.line}:${log.loc.column}) `;
		return prefix + log.message;
	};
} });

//#endregion
//#region src/utils/normalize-hook.ts
function normalizeHook(hook) {
	if (typeof hook === "function" || typeof hook === "string") return {
		handler: hook,
		options: {},
		meta: {}
	};
	if (typeof hook === "object" && hook !== null) {
		const { handler, order,...options } = hook;
		return {
			handler,
			options,
			meta: { order }
		};
	}
	unreachable("Invalid hook type");
}
var init_normalize_hook = __esm({ "src/utils/normalize-hook.ts"() {
	init_misc();
} });

//#endregion
//#region src/constants/plugin.ts
var ENUMERATED_INPUT_PLUGIN_HOOK_NAMES, ENUMERATED_OUTPUT_PLUGIN_HOOK_NAMES, ENUMERATED_PLUGIN_HOOK_NAMES, DEFINED_HOOK_NAMES;
var init_plugin$2 = __esm({ "src/constants/plugin.ts"() {
	ENUMERATED_INPUT_PLUGIN_HOOK_NAMES = [
		"options",
		"buildStart",
		"resolveId",
		"load",
		"transform",
		"moduleParsed",
		"buildEnd",
		"onLog",
		"resolveDynamicImport",
		"closeBundle",
		"closeWatcher",
		"watchChange"
	];
	ENUMERATED_OUTPUT_PLUGIN_HOOK_NAMES = [
		"augmentChunkHash",
		"outputOptions",
		"renderChunk",
		"renderStart",
		"renderError",
		"writeBundle",
		"generateBundle"
	];
	ENUMERATED_PLUGIN_HOOK_NAMES = [
		...ENUMERATED_INPUT_PLUGIN_HOOK_NAMES,
		...ENUMERATED_OUTPUT_PLUGIN_HOOK_NAMES,
		"footer",
		"banner",
		"intro",
		"outro"
	];
	DEFINED_HOOK_NAMES = {
		[ENUMERATED_PLUGIN_HOOK_NAMES[0]]: ENUMERATED_PLUGIN_HOOK_NAMES[0],
		[ENUMERATED_PLUGIN_HOOK_NAMES[1]]: ENUMERATED_PLUGIN_HOOK_NAMES[1],
		[ENUMERATED_PLUGIN_HOOK_NAMES[2]]: ENUMERATED_PLUGIN_HOOK_NAMES[2],
		[ENUMERATED_PLUGIN_HOOK_NAMES[3]]: ENUMERATED_PLUGIN_HOOK_NAMES[3],
		[ENUMERATED_PLUGIN_HOOK_NAMES[4]]: ENUMERATED_PLUGIN_HOOK_NAMES[4],
		[ENUMERATED_PLUGIN_HOOK_NAMES[5]]: ENUMERATED_PLUGIN_HOOK_NAMES[5],
		[ENUMERATED_PLUGIN_HOOK_NAMES[6]]: ENUMERATED_PLUGIN_HOOK_NAMES[6],
		[ENUMERATED_PLUGIN_HOOK_NAMES[7]]: ENUMERATED_PLUGIN_HOOK_NAMES[7],
		[ENUMERATED_PLUGIN_HOOK_NAMES[8]]: ENUMERATED_PLUGIN_HOOK_NAMES[8],
		[ENUMERATED_PLUGIN_HOOK_NAMES[9]]: ENUMERATED_PLUGIN_HOOK_NAMES[9],
		[ENUMERATED_PLUGIN_HOOK_NAMES[10]]: ENUMERATED_PLUGIN_HOOK_NAMES[10],
		[ENUMERATED_PLUGIN_HOOK_NAMES[11]]: ENUMERATED_PLUGIN_HOOK_NAMES[11],
		[ENUMERATED_PLUGIN_HOOK_NAMES[12]]: ENUMERATED_PLUGIN_HOOK_NAMES[12],
		[ENUMERATED_PLUGIN_HOOK_NAMES[13]]: ENUMERATED_PLUGIN_HOOK_NAMES[13],
		[ENUMERATED_PLUGIN_HOOK_NAMES[14]]: ENUMERATED_PLUGIN_HOOK_NAMES[14],
		[ENUMERATED_PLUGIN_HOOK_NAMES[15]]: ENUMERATED_PLUGIN_HOOK_NAMES[15],
		[ENUMERATED_PLUGIN_HOOK_NAMES[16]]: ENUMERATED_PLUGIN_HOOK_NAMES[16],
		[ENUMERATED_PLUGIN_HOOK_NAMES[17]]: ENUMERATED_PLUGIN_HOOK_NAMES[17],
		[ENUMERATED_PLUGIN_HOOK_NAMES[18]]: ENUMERATED_PLUGIN_HOOK_NAMES[18],
		[ENUMERATED_PLUGIN_HOOK_NAMES[19]]: ENUMERATED_PLUGIN_HOOK_NAMES[19],
		[ENUMERATED_PLUGIN_HOOK_NAMES[20]]: ENUMERATED_PLUGIN_HOOK_NAMES[20],
		[ENUMERATED_PLUGIN_HOOK_NAMES[21]]: ENUMERATED_PLUGIN_HOOK_NAMES[21],
		[ENUMERATED_PLUGIN_HOOK_NAMES[22]]: ENUMERATED_PLUGIN_HOOK_NAMES[22]
	};
} });

//#endregion
//#region src/utils/async-flatten.ts
async function asyncFlatten(array) {
	do
		array = (await Promise.all(array)).flat(Infinity);
	while (array.some((v$1) => v$1?.then));
	return array;
}
var init_async_flatten = __esm({ "src/utils/async-flatten.ts"() {} });

//#endregion
//#region src/utils/normalize-plugin-option.ts
function checkOutputPluginOption(plugins, onLog) {
	for (const plugin of plugins) for (const hook of ENUMERATED_INPUT_PLUGIN_HOOK_NAMES) if (hook in plugin) {
		delete plugin[hook];
		onLog(LOG_LEVEL_WARN, logInputHookInOutputPlugin(plugin.name, hook));
	}
	return plugins;
}
function normalizePlugins(plugins, anonymousPrefix) {
	for (const [index, plugin] of plugins.entries()) {
		if ("_parallel" in plugin) continue;
		if (plugin instanceof BuiltinPlugin) continue;
		if (!plugin.name) plugin.name = `${anonymousPrefix}${index + 1}`;
	}
	return plugins;
}
var normalizePluginOption, ANONYMOUS_PLUGIN_PREFIX, ANONYMOUS_OUTPUT_PLUGIN_PREFIX;
var init_normalize_plugin_option = __esm({ "src/utils/normalize-plugin-option.ts"() {
	init_constructors();
	init_plugin$2();
	init_logging();
	init_logs();
	init_async_flatten();
	normalizePluginOption = async (plugins) => (await asyncFlatten([plugins])).filter(Boolean);
	ANONYMOUS_PLUGIN_PREFIX = "at position ";
	ANONYMOUS_OUTPUT_PLUGIN_PREFIX = "at output position ";
} });

//#endregion
//#region src/plugin/minimal-plugin-context.ts
var MinimalPluginContextImpl;
var init_minimal_plugin_context = __esm({ "src/plugin/minimal-plugin-context.ts"() {
	init_src();
	init_log_handler();
	init_logging();
	init_logs();
	MinimalPluginContextImpl = class {
		info;
		warn;
		debug;
		meta;
		constructor(onLog, logLevel, pluginName, watchMode, hookName) {
			this.pluginName = pluginName;
			this.hookName = hookName;
			this.debug = getLogHandler(LOG_LEVEL_DEBUG, "PLUGIN_LOG", onLog, pluginName, logLevel);
			this.info = getLogHandler(LOG_LEVEL_INFO, "PLUGIN_LOG", onLog, pluginName, logLevel);
			this.warn = getLogHandler(LOG_LEVEL_WARN, "PLUGIN_WARNING", onLog, pluginName, logLevel);
			this.meta = {
				rollupVersion: "4.23.0",
				rolldownVersion: VERSION,
				watchMode
			};
		}
		error(e$5) {
			return error(logPluginError(normalizeLog(e$5), this.pluginName, { hook: this.hookName }));
		}
	};
} });

//#endregion
//#region src/plugin/plugin-driver.ts
function getObjectPlugins(plugins) {
	return plugins.filter((plugin) => {
		if (!plugin) return void 0;
		if ("_parallel" in plugin) return void 0;
		if (plugin instanceof BuiltinPlugin) return void 0;
		return plugin;
	});
}
function getSortedPlugins(hookName, plugins) {
	const pre = [];
	const normal = [];
	const post = [];
	for (const plugin of plugins) {
		const hook = plugin[hookName];
		if (hook) {
			if (typeof hook === "object") {
				if (hook.order === "pre") {
					pre.push(plugin);
					continue;
				}
				if (hook.order === "post") {
					post.push(plugin);
					continue;
				}
			}
			normal.push(plugin);
		}
	}
	return [
		...pre,
		...normal,
		...post
	];
}
var PluginDriver;
var init_plugin_driver = __esm({ "src/plugin/plugin-driver.ts"() {
	init_constructors();
	init_logger();
	init_logging();
	init_normalize_hook();
	init_normalize_plugin_option();
	init_minimal_plugin_context();
	PluginDriver = class {
		static async callOptionsHook(inputOptions, watchMode = false) {
			const logLevel = inputOptions.logLevel || LOG_LEVEL_INFO;
			const plugins = getSortedPlugins("options", getObjectPlugins(await normalizePluginOption(inputOptions.plugins)));
			const logger = getLogger(plugins, getOnLog(inputOptions, logLevel), logLevel, watchMode);
			for (const plugin of plugins) {
				const name = plugin.name || "unknown";
				const options = plugin.options;
				if (options) {
					const { handler } = normalizeHook(options);
					const result = await handler.call(new MinimalPluginContextImpl(logger, logLevel, name, watchMode, "onLog"), inputOptions);
					if (result) inputOptions = result;
				}
			}
			return inputOptions;
		}
		static callOutputOptionsHook(rawPlugins, outputOptions, onLog, logLevel, watchMode) {
			const sortedPlugins = getSortedPlugins("outputOptions", getObjectPlugins(rawPlugins));
			for (const plugin of sortedPlugins) {
				const name = plugin.name || "unknown";
				const options = plugin.outputOptions;
				if (options) {
					const { handler } = normalizeHook(options);
					const result = handler.call(new MinimalPluginContextImpl(onLog, logLevel, name, watchMode), outputOptions);
					if (result) outputOptions = result;
				}
			}
			return outputOptions;
		}
	};
} });

//#endregion
//#region src/utils/validator.ts
function validateCliOptions(options) {
	let parsed = v.safeParse(CliOptionsSchema, options);
	return [parsed.output, parsed.issues?.map((issue) => {
		const option = issue.path?.map((pathItem) => pathItem.key).join(" ");
		return `Invalid value for option ${option}: ${issue.message}`;
	})];
}
function validateOption(key, options) {
	if (globalThis.process?.env?.ROLLDOWN_OPTIONS_VALIDATION === "loose") return;
	let parsed = v.safeParse(key === "input" ? InputOptionsSchema : OutputOptionsSchema, options);
	if (!parsed.success) {
		const errors = parsed.issues.map((issue) => {
			const issuePaths = issue.path.map((path$1) => path$1.key);
			let issueMsg = issue.message;
			if (issue.type === "union") {
				const subIssue = issue.issues?.find((i$22) => !(i$22.type !== issue.received && i$22.input === issue.input));
				if (subIssue) {
					if (subIssue.path) issuePaths.push(subIssue.path.map((path$1) => path$1.key));
					issueMsg = subIssue.message;
				}
			}
			const stringPath = issuePaths.join(".");
			const helper = key === "input" ? inputHelperMsgRecord[stringPath] : outputHelperMsgRecord[stringPath];
			if (helper && helper.ignored) return "";
			return `- For the "${stringPath}". ${issueMsg}. ${helper ? helper.msg : ""}`;
		}).filter(Boolean);
		if (errors.length) throw new Error(`Failed validate ${key} options.\n` + errors.join("\n"));
	}
}
function getInputCliKeys() {
	return v.keyof(InputCliOptionsSchema).options;
}
function getOutputCliKeys() {
	return v.keyof(OutputCliOptionsSchema).options;
}
function getJsonSchema() {
	return toJsonSchema(CliOptionsSchema);
}
var StringOrRegExpSchema, LogLevelSchema, LogLevelOptionSchema, LogLevelWithErrorSchema, RollupLogSchema, RollupLogWithStringSchema, InputOptionSchema, ExternalSchema, ModuleTypesSchema, JsxOptionsSchema, HelperModeSchema, DecoratorOptionSchema, HelpersSchema, RewriteImportExtensionsSchema, TypescriptSchema, AssumptionsSchema, TransformOptionsSchema, WatchOptionsSchema, ChecksOptionsSchema, MinifyOptionsSchema, ResolveOptionsSchema, TreeshakingOptionsSchema, OnLogSchema, OnwarnSchema, HmrSchema, InputOptionsSchema, InputCliOverrideSchema, InputCliOptionsSchema, ESTarget, ModuleFormatSchema, AddonFunctionSchema, ChunkFileNamesSchema, AssetFileNamesSchema, SanitizeFileNameSchema, GlobalsFunctionSchema, AdvancedChunksSchema, OutputOptionsSchema, getAddonDescription, OutputCliOverrideSchema, OutputCliOptionsSchema, CliOptionsSchema, inputHelperMsgRecord, outputHelperMsgRecord;
var init_validator = __esm({ "src/utils/validator.ts"() {
	StringOrRegExpSchema = v.union([v.string(), v.instance(RegExp)]);
	LogLevelSchema = v.union([
		v.literal("debug"),
		v.literal("info"),
		v.literal("warn")
	]);
	LogLevelOptionSchema = v.union([LogLevelSchema, v.literal("silent")]);
	LogLevelWithErrorSchema = v.union([LogLevelSchema, v.literal("error")]);
	RollupLogSchema = v.any();
	RollupLogWithStringSchema = v.union([RollupLogSchema, v.string()]);
	InputOptionSchema = v.union([
		v.string(),
		v.array(v.string()),
		v.record(v.string(), v.string())
	]);
	ExternalSchema = v.union([
		StringOrRegExpSchema,
		v.array(StringOrRegExpSchema),
		v.pipe(v.function(), v.args(v.tuple([
			v.string(),
			v.optional(v.string()),
			v.boolean()
		])), v.returns(v.nullish(v.boolean())))
	]);
	ModuleTypesSchema = v.record(v.string(), v.union([
		v.literal("base64"),
		v.literal("binary"),
		v.literal("css"),
		v.literal("dataurl"),
		v.literal("empty"),
		v.literal("js"),
		v.literal("json"),
		v.literal("jsx"),
		v.literal("text"),
		v.literal("ts"),
		v.literal("tsx")
	]));
	JsxOptionsSchema = v.strictObject({
		development: v.pipe(v.optional(v.boolean()), v.description("Development specific information")),
		factory: v.pipe(v.optional(v.string()), v.description("Jsx element transformation")),
		fragment: v.pipe(v.optional(v.string()), v.description("Jsx fragment transformation")),
		importSource: v.pipe(v.optional(v.string()), v.description("Import the factory of element and fragment if mode is classic")),
		jsxImportSource: v.pipe(v.optional(v.string()), v.description("Import the factory of element and fragment if mode is automatic")),
		mode: v.pipe(v.optional(v.union([
			v.literal("classic"),
			v.literal("automatic"),
			v.literal("preserve")
		])), v.description("Jsx transformation mode")),
		refresh: v.pipe(v.optional(v.boolean()), v.description("React refresh transformation"))
	});
	HelperModeSchema = v.union([v.literal("Runtime"), v.literal("External")]);
	DecoratorOptionSchema = v.object({
		legacy: v.optional(v.boolean()),
		emitDecoratorMetadata: v.optional(v.boolean())
	});
	HelpersSchema = v.object({ mode: v.optional(HelperModeSchema) });
	RewriteImportExtensionsSchema = v.union([
		v.literal("rewrite"),
		v.literal("remove"),
		v.boolean()
	]);
	TypescriptSchema = v.object({
		jsxPragma: v.optional(v.string()),
		jsxPragmaFrag: v.optional(v.string()),
		onlyRemoveTypeImports: v.optional(v.boolean()),
		allowNamespaces: v.optional(v.boolean()),
		allowDeclareFields: v.optional(v.boolean()),
		declaration: v.optional(v.object({
			stripInternal: v.optional(v.boolean()),
			sourcemap: v.optional(v.boolean())
		})),
		rewriteImportExtensions: v.optional(RewriteImportExtensionsSchema)
	});
	AssumptionsSchema = v.object({
		ignoreFunctionLength: v.optional(v.boolean()),
		noDocumentAll: v.optional(v.boolean()),
		objectRestNoSymbols: v.optional(v.boolean()),
		pureGetters: v.optional(v.boolean()),
		setPublicClassFields: v.optional(v.boolean())
	});
	TransformOptionsSchema = v.object({
		assumptions: v.optional(AssumptionsSchema),
		typescript: v.optional(TypescriptSchema),
		helpers: v.optional(HelpersSchema),
		decorators: v.optional(DecoratorOptionSchema)
	});
	WatchOptionsSchema = v.strictObject({
		chokidar: v.optional(v.never(`The "watch.chokidar" option is deprecated, please use "watch.notify" instead of it`)),
		exclude: v.optional(v.union([StringOrRegExpSchema, v.array(StringOrRegExpSchema)])),
		include: v.optional(v.union([StringOrRegExpSchema, v.array(StringOrRegExpSchema)])),
		notify: v.pipe(v.optional(v.strictObject({
			compareContents: v.optional(v.boolean()),
			pollInterval: v.optional(v.number())
		})), v.description("Notify options")),
		skipWrite: v.pipe(v.optional(v.boolean()), v.description("Skip the bundle.write() step")),
		buildDelay: v.pipe(v.optional(v.number()), v.description("Throttle watch rebuilds"))
	});
	ChecksOptionsSchema = v.strictObject({
		circularDependency: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting circular dependency")),
		eval: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting eval")),
		missingGlobalName: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting missing global name")),
		missingNameOptionForIifeExport: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting missing name option for iife export")),
		mixedExport: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting mixed export")),
		unresolvedEntry: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting unresolved entry")),
		unresolvedImport: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting unresolved import")),
		filenameConflict: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting filename conflict")),
		commonJsVariableInEsm: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting common js variable in esm")),
		importIsUndefined: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting import is undefined")),
		configurationFieldConflict: v.pipe(v.optional(v.boolean()), v.description("Whether to emit warning when detecting configuration field conflict"))
	});
	MinifyOptionsSchema = v.strictObject({
		mangle: v.optional(v.boolean()),
		compress: v.optional(v.boolean()),
		removeWhitespace: v.optional(v.boolean())
	});
	ResolveOptionsSchema = v.strictObject({
		alias: v.optional(v.record(v.string(), v.union([v.string(), v.array(v.string())]))),
		aliasFields: v.optional(v.array(v.array(v.string()))),
		conditionNames: v.optional(v.array(v.string())),
		extensionAlias: v.optional(v.record(v.string(), v.array(v.string()))),
		exportsFields: v.optional(v.array(v.array(v.string()))),
		extensions: v.optional(v.array(v.string())),
		mainFields: v.optional(v.array(v.string())),
		mainFiles: v.optional(v.array(v.string())),
		modules: v.optional(v.array(v.string())),
		symlinks: v.optional(v.boolean()),
		tsconfigFilename: v.optional(v.string())
	});
	TreeshakingOptionsSchema = v.union([v.boolean(), v.looseObject({
		annotations: v.optional(v.boolean()),
		manualPureFunctions: v.optional(v.array(v.string())),
		unknownGlobalSideEffects: v.optional(v.boolean())
	})]);
	OnLogSchema = v.pipe(v.function(), v.args(v.tuple([
		LogLevelSchema,
		RollupLogSchema,
		v.pipe(v.function(), v.args(v.tuple([LogLevelWithErrorSchema, RollupLogWithStringSchema])))
	])));
	OnwarnSchema = v.pipe(v.function(), v.args(v.tuple([RollupLogSchema, v.pipe(v.function(), v.args(v.tuple([v.union([RollupLogWithStringSchema, v.pipe(v.function(), v.returns(RollupLogWithStringSchema))])])))])));
	HmrSchema = v.union([v.boolean(), v.strictObject({
		port: v.optional(v.number()),
		host: v.optional(v.string()),
		implement: v.optional(v.string())
	})]);
	InputOptionsSchema = v.strictObject({
		input: v.optional(InputOptionSchema),
		plugins: v.optional(v.custom(() => true)),
		external: v.optional(ExternalSchema),
		resolve: v.optional(ResolveOptionsSchema),
		cwd: v.pipe(v.optional(v.string()), v.description("Current working directory")),
		platform: v.pipe(v.optional(v.union([
			v.literal("browser"),
			v.literal("neutral"),
			v.literal("node")
		])), v.description(`Platform for which the code should be generated (node, ${colors.underline("browser")}, neutral)`)),
		shimMissingExports: v.pipe(v.optional(v.boolean()), v.description("Create shim variables for missing exports")),
		treeshake: v.optional(TreeshakingOptionsSchema),
		logLevel: v.pipe(v.optional(LogLevelOptionSchema), v.description(`Log level (${colors.dim("silent")}, ${colors.underline(colors.gray("info"))}, debug, ${colors.yellow("warn")})`)),
		onLog: v.optional(OnLogSchema),
		onwarn: v.optional(OnwarnSchema),
		moduleTypes: v.pipe(v.optional(ModuleTypesSchema), v.description("Module types for customized extensions")),
		experimental: v.optional(v.strictObject({
			disableLiveBindings: v.optional(v.boolean()),
			enableComposingJsPlugins: v.optional(v.boolean()),
			resolveNewUrlToAsset: v.optional(v.boolean()),
			strictExecutionOrder: v.optional(v.boolean()),
			hmr: v.optional(HmrSchema)
		})),
		define: v.pipe(v.optional(v.record(v.string(), v.string())), v.description("Define global variables")),
		inject: v.optional(v.record(v.string(), v.union([v.string(), v.tuple([v.string(), v.string()])]))),
		profilerNames: v.optional(v.boolean()),
		jsx: v.optional(v.union([
			v.boolean(),
			JsxOptionsSchema,
			v.string("react"),
			v.string("react-jsx"),
			v.string("preserve")
		])),
		transform: v.optional(TransformOptionsSchema),
		watch: v.optional(v.union([WatchOptionsSchema, v.literal(false)])),
		dropLabels: v.pipe(v.optional(v.array(v.string())), v.description("Remove labeled statements with these label names")),
		checks: v.optional(ChecksOptionsSchema),
		keepNames: v.pipe(v.optional(v.boolean()), v.description("Keep function/class name")),
		debug: v.pipe(v.optional(v.object({ sessionId: v.pipe(v.optional(v.string()), v.description("Used to name the build.")) })), v.description("Enable debug mode. Emit debug information to disk. This might slow down the build process significantly."))
	});
	InputCliOverrideSchema = v.strictObject({
		input: v.pipe(v.optional(v.array(v.string())), v.description("Entry file")),
		external: v.pipe(v.optional(v.array(v.string())), v.description("Comma-separated list of module ids to exclude from the bundle `<module-id>,...`")),
		inject: v.pipe(v.optional(v.record(v.string(), v.string())), v.description("Inject import statements on demand")),
		treeshake: v.pipe(v.optional(v.boolean()), v.description("enable treeshaking")),
		jsx: v.pipe(v.optional(JsxOptionsSchema), v.description("enable jsx"))
	});
	InputCliOptionsSchema = v.omit(v.strictObject({
		...InputOptionsSchema.entries,
		...InputCliOverrideSchema.entries
	}), [
		"plugins",
		"onwarn",
		"onLog",
		"resolve",
		"experimental",
		"profilerNames",
		"watch"
	]);
	ESTarget = /* @__PURE__ */ function(ESTarget$1) {
		ESTarget$1["ES6"] = "es6";
		ESTarget$1["ES2015"] = "es2015";
		ESTarget$1["ES2016"] = "es2016";
		ESTarget$1["ES2017"] = "es2017";
		ESTarget$1["ES2018"] = "es2018";
		ESTarget$1["ES2019"] = "es2019";
		ESTarget$1["ES2020"] = "es2020";
		ESTarget$1["ES2021"] = "es2021";
		ESTarget$1["ES2022"] = "es2022";
		ESTarget$1["ES2023"] = "es2023";
		ESTarget$1["ES2024"] = "es2024";
		ESTarget$1["ESNext"] = "esnext";
		return ESTarget$1;
	}(ESTarget || {});
	ModuleFormatSchema = v.union([
		v.literal("es"),
		v.literal("cjs"),
		v.literal("esm"),
		v.literal("module"),
		v.literal("commonjs"),
		v.literal("iife"),
		v.literal("umd")
	]);
	AddonFunctionSchema = v.pipe(v.function(), v.args(v.tuple([v.custom(() => true)])), v.returnsAsync(v.unionAsync([v.string(), v.pipeAsync(v.promise(), v.awaitAsync(), v.string())])));
	ChunkFileNamesSchema = v.union([v.string(), v.pipe(v.function(), v.args(v.tuple([v.custom(() => true)])), v.returns(v.string()))]);
	AssetFileNamesSchema = v.union([v.string(), v.pipe(v.function(), v.args(v.tuple([v.custom(() => true)])), v.returns(v.string()))]);
	SanitizeFileNameSchema = v.union([v.boolean(), v.pipe(v.function(), v.args(v.tuple([v.string()])), v.returns(v.string()))]);
	GlobalsFunctionSchema = v.pipe(v.function(), v.args(v.tuple([v.string()])), v.returns(v.string()));
	AdvancedChunksSchema = v.strictObject({
		minSize: v.optional(v.number()),
		maxSize: v.optional(v.number()),
		minModuleSize: v.optional(v.number()),
		maxModuleSize: v.optional(v.number()),
		minShareCount: v.optional(v.number()),
		groups: v.optional(v.array(v.strictObject({
			name: v.string(),
			test: v.optional(v.union([v.string(), v.instance(RegExp)])),
			priority: v.optional(v.number()),
			minSize: v.optional(v.number()),
			minShareCount: v.optional(v.number()),
			maxSize: v.optional(v.number()),
			minModuleSize: v.optional(v.number()),
			maxModuleSize: v.optional(v.number())
		})))
	});
	OutputOptionsSchema = v.strictObject({
		dir: v.pipe(v.optional(v.string()), v.description("Output directory, defaults to `dist` if `file` is not set")),
		file: v.pipe(v.optional(v.string()), v.description("Single output file")),
		exports: v.pipe(v.optional(v.union([
			v.literal("auto"),
			v.literal("named"),
			v.literal("default"),
			v.literal("none")
		])), v.description(`Specify a export mode (${colors.underline("auto")}, named, default, none)`)),
		hashCharacters: v.pipe(v.optional(v.union([
			v.literal("base64"),
			v.literal("base36"),
			v.literal("hex")
		])), v.description("Use the specified character set for file hashes")),
		format: v.pipe(v.optional(ModuleFormatSchema), v.description(`Output format of the generated bundle (supports ${colors.underline("esm")}, cjs, and iife)`)),
		sourcemap: v.pipe(v.optional(v.union([
			v.boolean(),
			v.literal("inline"),
			v.literal("hidden")
		])), v.description(`Generate sourcemap (\`-s inline\` for inline, or ${colors.bold("pass the `-s` on the last argument if you want to generate `.map` file")})`)),
		sourcemapDebugIds: v.pipe(v.optional(v.boolean()), v.description("Inject sourcemap debug IDs")),
		sourcemapIgnoreList: v.optional(v.union([v.boolean(), v.custom(() => true)])),
		sourcemapPathTransform: v.optional(v.custom(() => true)),
		banner: v.optional(v.union([v.string(), AddonFunctionSchema])),
		footer: v.optional(v.union([v.string(), AddonFunctionSchema])),
		intro: v.optional(v.union([v.string(), AddonFunctionSchema])),
		outro: v.optional(v.union([v.string(), AddonFunctionSchema])),
		extend: v.pipe(v.optional(v.boolean()), v.description("Extend global variable defined by name in IIFE / UMD formats")),
		esModule: v.optional(v.union([v.boolean(), v.literal("if-default-prop")])),
		assetFileNames: v.optional(AssetFileNamesSchema),
		entryFileNames: v.optional(ChunkFileNamesSchema),
		chunkFileNames: v.optional(ChunkFileNamesSchema),
		cssEntryFileNames: v.optional(ChunkFileNamesSchema),
		cssChunkFileNames: v.optional(ChunkFileNamesSchema),
		sanitizeFileName: v.optional(SanitizeFileNameSchema),
		minify: v.pipe(v.optional(v.union([
			v.boolean(),
			v.string("dce-only"),
			MinifyOptionsSchema
		])), v.description("Minify the bundled file")),
		name: v.pipe(v.optional(v.string()), v.description("Name for UMD / IIFE format outputs")),
		globals: v.pipe(v.optional(v.union([v.record(v.string(), v.string()), GlobalsFunctionSchema])), v.description("Global variable of UMD / IIFE dependencies (syntax: `key=value`)")),
		externalLiveBindings: v.pipe(v.optional(v.boolean()), v.description("external live bindings")),
		inlineDynamicImports: v.pipe(v.optional(v.boolean()), v.description("Inline dynamic imports")),
		advancedChunks: v.optional(AdvancedChunksSchema),
		comments: v.pipe(v.optional(v.union([v.literal("none"), v.literal("preserve-legal")])), v.description("Control comments in the output")),
		plugins: v.optional(v.custom(() => true)),
		polyfillRequire: v.pipe(v.optional(v.boolean()), v.description("Disable require polyfill injection")),
		target: v.pipe(v.optional(v.enum(ESTarget)), v.description("The JavaScript target environment")),
		hoistTransitiveImports: v.optional(v.custom((input) => {
			if (input) return false;
			return true;
		}, () => `The 'true' value is not supported`))
	});
	getAddonDescription = (placement, wrapper) => {
		return `Code to insert the ${colors.bold(placement)} of the bundled file (${colors.bold(wrapper)} the wrapper function)`;
	};
	OutputCliOverrideSchema = v.strictObject({
		assetFileNames: v.pipe(v.optional(v.string()), v.description("Name pattern for asset files")),
		entryFileNames: v.pipe(v.optional(v.string()), v.description("Name pattern for emitted entry chunks")),
		chunkFileNames: v.pipe(v.optional(v.string()), v.description("Name pattern for emitted secondary chunks")),
		cssEntryFileNames: v.pipe(v.optional(v.string()), v.description("Name pattern for emitted css entry chunks")),
		cssChunkFileNames: v.pipe(v.optional(v.string()), v.description("Name pattern for emitted css secondary chunks")),
		sanitizeFileName: v.pipe(v.optional(v.boolean()), v.description("Sanitize file name")),
		banner: v.pipe(v.optional(v.string()), v.description(getAddonDescription("top", "outside"))),
		footer: v.pipe(v.optional(v.string()), v.description(getAddonDescription("bottom", "outside"))),
		intro: v.pipe(v.optional(v.string()), v.description(getAddonDescription("top", "inside"))),
		outro: v.pipe(v.optional(v.string()), v.description(getAddonDescription("bottom", "inside"))),
		esModule: v.pipe(v.optional(v.boolean()), v.description("Always generate `__esModule` marks in non-ESM formats, defaults to `if-default-prop` (use `--no-esModule` to always disable)")),
		globals: v.pipe(v.optional(v.record(v.string(), v.string())), v.description("Global variable of UMD / IIFE dependencies (syntax: `key=value`)")),
		advancedChunks: v.pipe(v.optional(v.strictObject({
			minSize: v.pipe(v.optional(v.number()), v.description("Minimum size of the chunk")),
			minShareCount: v.pipe(v.optional(v.number()), v.description("Minimum share count of the chunk"))
		})), v.description("Global variable of UMD / IIFE dependencies (syntax: `key=value`)")),
		minify: v.pipe(v.optional(v.boolean()), v.description("Minify the bundled file"))
	});
	OutputCliOptionsSchema = v.omit(v.strictObject({
		...OutputOptionsSchema.entries,
		...OutputCliOverrideSchema.entries
	}), [
		"sourcemapIgnoreList",
		"sourcemapPathTransform",
		"plugins",
		"hoistTransitiveImports"
	]);
	CliOptionsSchema = v.strictObject({
		config: v.pipe(v.optional(v.union([v.string(), v.boolean()])), v.description("Path to the config file (default: `rolldown.config.js`)")),
		help: v.pipe(v.optional(v.boolean()), v.description("Show help")),
		version: v.pipe(v.optional(v.boolean()), v.description("Show version number")),
		watch: v.pipe(v.optional(v.boolean()), v.description("Watch files in bundle and rebuild on changes")),
		...InputCliOptionsSchema.entries,
		...OutputCliOptionsSchema.entries
	});
	inputHelperMsgRecord = { output: { ignored: true } };
	outputHelperMsgRecord = {};
} });

//#endregion
//#region src/constants/plugin-context.ts
var SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF;
var init_plugin_context$1 = __esm({ "src/constants/plugin-context.ts"() {
	SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF = Symbol("plugin-context-resolve-caller");
} });

//#endregion
//#region src/options/normalized-input-options.ts
var NormalizedInputOptionsImpl;
var init_normalized_input_options = __esm({ "src/options/normalized-input-options.ts"() {
	NormalizedInputOptionsImpl = class {
		inner;
		constructor(inner, onLog) {
			this.onLog = onLog;
			this.inner = inner;
		}
		get shimMissingExports() {
			return this.inner.shimMissingExports;
		}
		get input() {
			return this.inner.input;
		}
		get cwd() {
			return this.inner.cwd ?? void 0;
		}
		get platform() {
			return this.inner.platform;
		}
	};
} });

//#endregion
//#region src/types/sourcemap.ts
function bindingifySourcemap(map) {
	if (map == null) return;
	return { inner: typeof map === "string" ? map : {
		file: map.file ?? void 0,
		mappings: map.mappings,
		sourceRoot: "sourceRoot" in map ? map.sourceRoot ?? void 0 : void 0,
		sources: map.sources?.map((s$10) => s$10 ?? void 0),
		sourcesContent: map.sourcesContent?.map((s$10) => s$10 ?? void 0),
		names: map.names,
		x_google_ignoreList: map.x_google_ignoreList,
		debugId: "debugId" in map ? map.debugId : void 0
	} };
}
var init_sourcemap = __esm({ "src/types/sourcemap.ts"() {} });

//#endregion
//#region src/utils/error.ts
function normalizeErrors(rawErrors) {
	const errors = rawErrors.map((e$5) => e$5 instanceof Error ? e$5 : Object.assign(new Error(), {
		kind: e$5.kind,
		message: e$5.message,
		stack: void 0
	}));
	let summary = `Build failed with ${errors.length} error${errors.length < 2 ? "" : "s"}:\n`;
	for (let i$22 = 0; i$22 < errors.length; i$22++) {
		summary += "\n";
		if (i$22 >= 5) {
			summary += "...";
			break;
		}
		summary += getErrorMessage(errors[i$22]);
	}
	const wrapper = new Error(summary);
	Object.defineProperty(wrapper, "errors", {
		configurable: true,
		enumerable: true,
		get: () => errors,
		set: (value) => Object.defineProperty(wrapper, "errors", {
			configurable: true,
			enumerable: true,
			value
		})
	});
	return wrapper;
}
function getErrorMessage(e$5) {
	if (Object.hasOwn(e$5, "kind")) return e$5.message;
	let s$10 = "";
	if (e$5.plugin) s$10 += `[plugin ${e$5.plugin}]`;
	const id$1 = e$5.id ?? e$5.loc?.file;
	if (id$1) {
		s$10 += " " + id$1;
		if (e$5.loc) s$10 += `:${e$5.loc.line}:${e$5.loc.column}`;
	}
	if (s$10) s$10 += "\n";
	const message = `${e$5.name ?? "Error"}: ${e$5.message}`;
	s$10 += message;
	if (e$5.frame) s$10 = joinNewLine(s$10, e$5.frame);
	if (e$5.stack) s$10 = joinNewLine(s$10, e$5.stack.replace(message, ""));
	return s$10;
}
function joinNewLine(s1, s2) {
	return s1.replace(/\n+$/, "") + "\n" + s2.replace(/^\n+/, "");
}
var init_error = __esm({ "src/utils/error.ts"() {} });

//#endregion
//#region src/utils/transform-module-info.ts
function transformModuleInfo(info, option) {
	return {
		get ast() {
			return unsupported("ModuleInfo#ast");
		},
		get code() {
			return info.code;
		},
		id: info.id,
		importers: info.importers,
		dynamicImporters: info.dynamicImporters,
		importedIds: info.importedIds,
		dynamicallyImportedIds: info.dynamicallyImportedIds,
		exports: info.exports,
		isEntry: info.isEntry,
		...option
	};
}
var init_transform_module_info = __esm({ "src/utils/transform-module-info.ts"() {
	init_misc();
} });

//#endregion
//#region src/utils/transform-side-effects.ts
function bindingifySideEffects(sideEffects) {
	switch (sideEffects) {
		case true: return import_binding.BindingHookSideEffects.True;
		case false: return import_binding.BindingHookSideEffects.False;
		case "no-treeshake": return import_binding.BindingHookSideEffects.NoTreeshake;
		case null:
		case void 0: return void 0;
		default: throw new Error(`Unexpected side effects: ${sideEffects}`);
	}
}
var init_transform_side_effects = __esm({ "src/utils/transform-side-effects.ts"() {} });

//#endregion
//#region src/utils/transform-sourcemap.ts
function isEmptySourcemapFiled(array) {
	if (!array) return true;
	if (array.length === 0 || !array[0]) return true;
	return false;
}
function normalizeTransformHookSourcemap(id$1, originalCode, rawMap) {
	if (!rawMap) return;
	let map = typeof rawMap === "object" ? rawMap : JSON.parse(rawMap);
	if (isEmptySourcemapFiled(map.sourcesContent)) map.sourcesContent = [originalCode];
	if (isEmptySourcemapFiled(map.sources) || map.sources && map.sources.length === 1 && map.sources[0] !== id$1) map.sources = [id$1];
	return map;
}
var init_transform_sourcemap = __esm({ "src/utils/transform-sourcemap.ts"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ANXBDSUI.js
var init_chunk_ANXBDSUI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ANXBDSUI.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3GOCSNFN.js
var init_chunk_3GOCSNFN = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3GOCSNFN.js"() {
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-LFJW7BOT.js
var init_chunk_LFJW7BOT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-LFJW7BOT.js"() {
	init_chunk_3GOCSNFN();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-7ZI6JRPB.js
var init_chunk_7ZI6JRPB = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-7ZI6JRPB.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OXJMERKM.js
var init_chunk_OXJMERKM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OXJMERKM.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-D6FCK2GA.js
function u$13(o$12, n$7, a$15) {
	let t$14 = (r$12) => o$12(r$12, ...n$7);
	return a$15 === void 0 ? t$14 : Object.assign(t$14, {
		lazy: a$15,
		lazyArgs: n$7
	});
}
var init_chunk_D6FCK2GA = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-D6FCK2GA.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WIMGWYZL.js
function u$11(r$12, n$7, o$12) {
	let a$15 = r$12.length - n$7.length;
	if (a$15 === 0) return r$12(...n$7);
	if (a$15 === 1) return u$13(r$12, n$7, o$12);
	throw new Error("Wrong number of arguments");
}
var init_chunk_WIMGWYZL = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WIMGWYZL.js"() {
	init_chunk_D6FCK2GA();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BSLJB6JE.js
var init_chunk_BSLJB6JE = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BSLJB6JE.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NJXNQM3G.js
var init_chunk_NJXNQM3G = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NJXNQM3G.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6RKHJ2CP.js
var init_chunk_6RKHJ2CP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6RKHJ2CP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QDGUNRDA.js
var init_chunk_QDGUNRDA = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QDGUNRDA.js"() {
	init_chunk_D6FCK2GA();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MYLLMFC7.js
var init_chunk_MYLLMFC7 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MYLLMFC7.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DEVKGLTN.js
var o$11, c$3, i$21;
var init_chunk_DEVKGLTN = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DEVKGLTN.js"() {
	o$11 = [
		"	",
		`
`,
		"\v",
		"\f",
		"\r",
		" ",
		"",
		"\xA0",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		" ",
		"\u2028",
		"\u2029",
		" ",
		" ",
		"　",
		"﻿"
	], c$3 = new Set([
		"-",
		"_",
		...o$11
	]), i$21 = (r$12) => {
		let e$5 = [], t$14 = "", u$14 = () => {
			t$14.length > 0 && (e$5.push(t$14), t$14 = "");
		};
		for (let s$10 of r$12) {
			if (c$3.has(s$10)) {
				u$14();
				continue;
			}
			if (/[a-z]$/u.test(t$14) && /[A-Z]/u.test(s$10)) u$14();
			else if (/[A-Z][A-Z]$/u.test(t$14) && /[a-z]/u.test(s$10)) {
				let n$7 = t$14.slice(-1);
				t$14 = t$14.slice(0, -1), u$14(), t$14 = n$7;
			} else /\d$/u.test(t$14) !== /\d/u.test(s$10) && u$14();
			t$14 += s$10;
		}
		return u$14(), e$5;
	};
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-N4JUOEMS.js
var init_chunk_N4JUOEMS = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-N4JUOEMS.js"() {
	init_chunk_DEVKGLTN();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YRJ25UV2.js
var init_chunk_YRJ25UV2 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YRJ25UV2.js"() {
	init_chunk_DEVKGLTN();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-Q5ASJ5N7.js
var init_chunk_Q5ASJ5N7 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-Q5ASJ5N7.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WZOX4VKU.js
var init_chunk_WZOX4VKU = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WZOX4VKU.js"() {
	init_chunk_DEVKGLTN();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-Y3VKZ3P5.js
var init_chunk_Y3VKZ3P5 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-Y3VKZ3P5.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZRKG4NSC.js
var init_chunk_ZRKG4NSC = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZRKG4NSC.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QJLMYOTX.js
var init_chunk_QJLMYOTX = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QJLMYOTX.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DM52TTEP.js
var init_chunk_DM52TTEP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DM52TTEP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2P44HXVH.js
var init_chunk_2P44HXVH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2P44HXVH.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZJS5DNQW.js
var init_chunk_ZJS5DNQW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZJS5DNQW.js"() {
	init_chunk_ANXBDSUI();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DH3BPT6T.js
var init_chunk_DH3BPT6T = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DH3BPT6T.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OAMXQXGR.js
var init_chunk_OAMXQXGR = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OAMXQXGR.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P5WDBFN2.js
var init_chunk_P5WDBFN2 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P5WDBFN2.js"() {
	init_chunk_DH3BPT6T();
	init_chunk_OAMXQXGR();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EMIEIAAH.js
var init_chunk_EMIEIAAH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EMIEIAAH.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-AHDTEY6L.js
var init_chunk_AHDTEY6L = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-AHDTEY6L.js"() {
	init_chunk_P5WDBFN2();
	init_chunk_EMIEIAAH();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4UEQNEAO.js
var init_chunk_4UEQNEAO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4UEQNEAO.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T4PLMLCP.js
var init_chunk_T4PLMLCP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T4PLMLCP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SSDL7ATG.js
var init_chunk_SSDL7ATG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SSDL7ATG.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EVIH3PFY.js
var init_chunk_EVIH3PFY = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EVIH3PFY.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-RBODUO3Q.js
var init_chunk_RBODUO3Q = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-RBODUO3Q.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-S52RID4A.js
var init_chunk_S52RID4A = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-S52RID4A.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WWPMIW33.js
var init_chunk_WWPMIW33 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WWPMIW33.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-57KROWWS.js
var init_chunk_57KROWWS = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-57KROWWS.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-V6HCOU6T.js
var init_chunk_V6HCOU6T = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-V6HCOU6T.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ALS6JP7S.js
var init_chunk_ALS6JP7S = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ALS6JP7S.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QTQTP2VB.js
var init_chunk_QTQTP2VB = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QTQTP2VB.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NFFV4IQT.js
var init_chunk_NFFV4IQT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NFFV4IQT.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FDH4IRIM.js
var init_chunk_FDH4IRIM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FDH4IRIM.js"() {
	init_chunk_EMIEIAAH();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QEKOZYJ5.js
var init_chunk_QEKOZYJ5 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QEKOZYJ5.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YDIA5YQI.js
var init_chunk_YDIA5YQI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YDIA5YQI.js"() {
	init_chunk_QEKOZYJ5();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6OEKBHIX.js
var init_chunk_6OEKBHIX = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6OEKBHIX.js"() {
	init_chunk_QEKOZYJ5();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GDGEDZJG.js
var init_chunk_GDGEDZJG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GDGEDZJG.js"() {
	init_chunk_QEKOZYJ5();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XE3XIKTJ.js
var init_chunk_XE3XIKTJ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XE3XIKTJ.js"() {
	init_chunk_QEKOZYJ5();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HVJXDSOP.js
var init_chunk_HVJXDSOP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HVJXDSOP.js"() {
	init_chunk_QEKOZYJ5();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MSOX5OUI.js
var init_chunk_MSOX5OUI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MSOX5OUI.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-C4OZY4Z2.js
var init_chunk_C4OZY4Z2 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-C4OZY4Z2.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P2PQB7KO.js
var init_chunk_P2PQB7KO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P2PQB7KO.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FZHIMCK6.js
var init_chunk_FZHIMCK6 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FZHIMCK6.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UHZ33J57.js
var init_chunk_UHZ33J57 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UHZ33J57.js"() {
	init_chunk_FZHIMCK6();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6RL33UFT.js
var init_chunk_6RL33UFT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6RL33UFT.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YNNF733L.js
var init_chunk_YNNF733L = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YNNF733L.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4YLWJIJ6.js
var init_chunk_4YLWJIJ6 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4YLWJIJ6.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KVHF7QRD.js
var init_chunk_KVHF7QRD = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KVHF7QRD.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UA6DVSZ3.js
var init_chunk_UA6DVSZ3 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UA6DVSZ3.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-G5B2IDWB.js
var init_chunk_G5B2IDWB = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-G5B2IDWB.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3FKP6OOU.js
var init_chunk_3FKP6OOU = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3FKP6OOU.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VFSOOVKJ.js
var init_chunk_VFSOOVKJ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VFSOOVKJ.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-K3UJMX27.js
var init_chunk_K3UJMX27 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-K3UJMX27.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-LE6I3KC6.js
var init_chunk_LE6I3KC6 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-LE6I3KC6.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ENS7GPLZ.js
var init_chunk_ENS7GPLZ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ENS7GPLZ.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DSLWSGID.js
var init_chunk_DSLWSGID = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DSLWSGID.js"() {
	init_chunk_EMIEIAAH();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-R72GEKLP.js
var init_chunk_R72GEKLP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-R72GEKLP.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3IFJP4R5.js
function d$3(...r$12) {
	return u$11(i$19, r$12);
}
var i$19;
var init_chunk_3IFJP4R5 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3IFJP4R5.js"() {
	init_chunk_WIMGWYZL();
	i$19 = (r$12, t$14) => {
		let a$15 = [[], []];
		for (let [o$12, e$5] of r$12.entries()) t$14(e$5, o$12, r$12) ? a$15[0].push(e$5) : a$15[1].push(e$5);
		return a$15;
	};
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J4EKWFDW.js
var init_chunk_J4EKWFDW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J4EKWFDW.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PGMPBC5Q.js
var init_chunk_PGMPBC5Q = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PGMPBC5Q.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GGYFZTDW.js
var init_chunk_GGYFZTDW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GGYFZTDW.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EDOGCRPU.js
var init_chunk_EDOGCRPU = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-EDOGCRPU.js"() {
	init_chunk_3GOCSNFN();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4NRWDO7P.js
var init_chunk_4NRWDO7P = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4NRWDO7P.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-W6ZHPGFP.js
var init_chunk_W6ZHPGFP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-W6ZHPGFP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J3IRE4DI.js
var init_chunk_J3IRE4DI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J3IRE4DI.js"() {
	init_chunk_DH3BPT6T();
	init_chunk_EMIEIAAH();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-RZUYD7QY.js
var init_chunk_RZUYD7QY = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-RZUYD7QY.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-A3PVMI4K.js
var init_chunk_A3PVMI4K = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-A3PVMI4K.js"() {
	init_chunk_OAMXQXGR();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4ZFFLFWV.js
var init_chunk_4ZFFLFWV = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-4ZFFLFWV.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-C6IMN7SF.js
var init_chunk_C6IMN7SF = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-C6IMN7SF.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NS6ZBRLP.js
var init_chunk_NS6ZBRLP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NS6ZBRLP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FMPZ2CLX.js
var init_chunk_FMPZ2CLX = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FMPZ2CLX.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3ZJAREUD.js
var init_chunk_3ZJAREUD = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3ZJAREUD.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZXVA7VDE.js
var init_chunk_ZXVA7VDE = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZXVA7VDE.js"() {
	init_chunk_LFJW7BOT();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-567G5ZXL.js
var init_chunk_567G5ZXL = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-567G5ZXL.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KQRZQWDE.js
var init_chunk_KQRZQWDE = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KQRZQWDE.js"() {
	init_chunk_567G5ZXL();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3D3RWAVJ.js
var init_chunk_3D3RWAVJ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3D3RWAVJ.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-K2FFNW24.js
var init_chunk_K2FFNW24 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-K2FFNW24.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5S4PYKVY.js
var init_chunk_5S4PYKVY = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5S4PYKVY.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5WKPQX7L.js
var init_chunk_5WKPQX7L = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5WKPQX7L.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BZNENX2T.js
var init_chunk_BZNENX2T = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BZNENX2T.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PDQFB3TV.js
var init_chunk_PDQFB3TV = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PDQFB3TV.js"() {
	init_chunk_BZNENX2T();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XMLUDZIW.js
var init_chunk_XMLUDZIW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XMLUDZIW.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GMMLSO2N.js
var init_chunk_GMMLSO2N = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GMMLSO2N.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5NQBDF4H.js
function t$5(...n$7) {
	return u$11(Object.keys, n$7);
}
var init_chunk_5NQBDF4H = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5NQBDF4H.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PFSVCZNE.js
var init_chunk_PFSVCZNE = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PFSVCZNE.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VO5MRBXA.js
var init_chunk_VO5MRBXA = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VO5MRBXA.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XUX3ZEXI.js
var init_chunk_XUX3ZEXI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XUX3ZEXI.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KZIKCQ56.js
var init_chunk_KZIKCQ56 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KZIKCQ56.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P3DXEVTH.js
var init_chunk_P3DXEVTH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P3DXEVTH.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6GTAPB47.js
var init_chunk_6GTAPB47 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-6GTAPB47.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NMC53JVB.js
var init_chunk_NMC53JVB = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NMC53JVB.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PULGOXDA.js
var init_chunk_PULGOXDA = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PULGOXDA.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OLNQBNAJ.js
var init_chunk_OLNQBNAJ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OLNQBNAJ.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QOEIYQAG.js
var init_chunk_QOEIYQAG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QOEIYQAG.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SFZGYJFI.js
var init_chunk_SFZGYJFI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SFZGYJFI.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-JJZ7E4YG.js
var init_chunk_JJZ7E4YG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-JJZ7E4YG.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VCYTMP4D.js
var init_chunk_VCYTMP4D = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VCYTMP4D.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-CAZXBO45.js
var init_chunk_CAZXBO45 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-CAZXBO45.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ENOHV5LT.js
var init_chunk_ENOHV5LT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ENOHV5LT.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-U753ZCO5.js
var init_chunk_U753ZCO5 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-U753ZCO5.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5DU4ITSF.js
var init_chunk_5DU4ITSF = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-5DU4ITSF.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GK5I7C4J.js
var init_chunk_GK5I7C4J = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GK5I7C4J.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HV3WACXG.js
var init_chunk_HV3WACXG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HV3WACXG.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-JK3VNB42.js
var init_chunk_JK3VNB42 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-JK3VNB42.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T45O7BFY.js
var init_chunk_T45O7BFY = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T45O7BFY.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OP5ZF26D.js
var init_chunk_OP5ZF26D = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OP5ZF26D.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BO3LQZNF.js
var init_chunk_BO3LQZNF = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-BO3LQZNF.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-I3D2BSWJ.js
var init_chunk_I3D2BSWJ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-I3D2BSWJ.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-7QX4DO53.js
var init_chunk_7QX4DO53 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-7QX4DO53.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VMV5GVZ5.js
var init_chunk_VMV5GVZ5 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VMV5GVZ5.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OWH4IQQW.js
var init_chunk_OWH4IQQW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OWH4IQQW.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ETADWPSK.js
var init_chunk_ETADWPSK = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ETADWPSK.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OFZGPGDZ.js
var init_chunk_OFZGPGDZ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OFZGPGDZ.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HGKLN5KY.js
var init_chunk_HGKLN5KY = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HGKLN5KY.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-R7PILVSQ.js
var init_chunk_R7PILVSQ = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-R7PILVSQ.js"() {
	init_chunk_HGKLN5KY();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HJSE3ESO.js
var init_chunk_HJSE3ESO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HJSE3ESO.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-B6PG574O.js
var init_chunk_B6PG574O = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-B6PG574O.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ICBBHOCR.js
var init_chunk_ICBBHOCR = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ICBBHOCR.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DAE7EJQ2.js
var init_chunk_DAE7EJQ2 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-DAE7EJQ2.js"() {
	init_chunk_EMIEIAAH();
	init_chunk_OAMXQXGR();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XPCYQPKH.js
var init_chunk_XPCYQPKH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XPCYQPKH.js"() {
	init_chunk_ANXBDSUI();
	init_chunk_D6FCK2GA();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FRNNS7AX.js
var init_chunk_FRNNS7AX = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-FRNNS7AX.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QJOWZFYO.js
var init_chunk_QJOWZFYO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-QJOWZFYO.js"() {
	init_chunk_FZHIMCK6();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VIBSXWWU.js
var init_chunk_VIBSXWWU = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VIBSXWWU.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T4H4IOYC.js
var init_chunk_T4H4IOYC = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T4H4IOYC.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GPLTWAVR.js
var init_chunk_GPLTWAVR = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GPLTWAVR.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PIX5OHMW.js
var init_chunk_PIX5OHMW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PIX5OHMW.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VVM5DH6Z.js
var init_chunk_VVM5DH6Z = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VVM5DH6Z.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PVYOMZ3I.js
var init_chunk_PVYOMZ3I = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-PVYOMZ3I.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P6LAFGAN.js
var init_chunk_P6LAFGAN = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-P6LAFGAN.js"() {
	init_chunk_ANXBDSUI();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SGAFZVQH.js
var init_chunk_SGAFZVQH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-SGAFZVQH.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MQDP6CFS.js
var init_chunk_MQDP6CFS = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-MQDP6CFS.js"() {
	init_chunk_SGAFZVQH();
	init_chunk_ANXBDSUI();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UZ6BOIAH.js
var init_chunk_UZ6BOIAH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UZ6BOIAH.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KI5UAETW.js
var init_chunk_KI5UAETW = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-KI5UAETW.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GYH2VCL4.js
var init_chunk_GYH2VCL4 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GYH2VCL4.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-26ILFTOP.js
var init_chunk_26ILFTOP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-26ILFTOP.js"() {
	init_chunk_SGAFZVQH();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2KIKGHAO.js
var init_chunk_2KIKGHAO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2KIKGHAO.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YVMG2XEU.js
var init_chunk_YVMG2XEU = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-YVMG2XEU.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZP6YI2JT.js
var init_chunk_ZP6YI2JT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZP6YI2JT.js"() {
	init_chunk_ANXBDSUI();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OCJMJQKO.js
var init_chunk_OCJMJQKO = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OCJMJQKO.js"() {
	init_chunk_P5WDBFN2();
	init_chunk_EMIEIAAH();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J7R2OSHS.js
var init_chunk_J7R2OSHS = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-J7R2OSHS.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GIKF2ZNG.js
var init_chunk_GIKF2ZNG = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GIKF2ZNG.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XWBKJZIP.js
var init_chunk_XWBKJZIP = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XWBKJZIP.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XHPQVWZM.js
var init_chunk_XHPQVWZM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XHPQVWZM.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2OQBQB3V.js
var init_chunk_2OQBQB3V = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-2OQBQB3V.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-H4OTHZJB.js
var init_chunk_H4OTHZJB = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-H4OTHZJB.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XZ6COQKM.js
function a$13(n$7, t$14, e$5) {
	return n$7(e$5[0]) ? (r$12) => t$14(r$12, ...e$5) : t$14(...e$5);
}
function u$12(...n$7) {
	return a$13(s$8, o$9, n$7);
}
function o$9(n$7, ...t$14) {
	for (let [e$5, r$12] of t$14) if (e$5(n$7)) return r$12(n$7);
	throw new Error("conditional: data failed for all cases");
}
function s$8(n$7) {
	if (!Array.isArray(n$7)) return !1;
	let [t$14, e$5, ...r$12] = n$7;
	return typeof t$14 == "function" && t$14.length <= 1 && typeof e$5 == "function" && e$5.length <= 1 && r$12.length === 0;
}
function R(n$7 = d$11) {
	return [T$7, n$7];
}
var x$2, T$7, d$11;
var init_chunk_XZ6COQKM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-XZ6COQKM.js"() {
	x$2 = Object.assign(u$12, { defaultCase: R });
	T$7 = () => !0, d$11 = () => {};
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T5XG33UI.js
var init_chunk_T5XG33UI = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-T5XG33UI.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZTNFU7RH.js
var init_chunk_ZTNFU7RH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-ZTNFU7RH.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OIQJEOF7.js
var init_chunk_OIQJEOF7 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-OIQJEOF7.js"() {} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GKXRNLHM.js
var init_chunk_GKXRNLHM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-GKXRNLHM.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NYIWN625.js
var init_chunk_NYIWN625 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-NYIWN625.js"() {
	init_chunk_LFJW7BOT();
	init_chunk_ANXBDSUI();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WPTI67A4.js
var init_chunk_WPTI67A4 = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WPTI67A4.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-W2ARC73P.js
var init_chunk_W2ARC73P = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-W2ARC73P.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3UBK2BVM.js
var init_chunk_3UBK2BVM = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-3UBK2BVM.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VFECZ57D.js
var init_chunk_VFECZ57D = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VFECZ57D.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VG2NVNXT.js
var init_chunk_VG2NVNXT = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-VG2NVNXT.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HJSE36CH.js
var init_chunk_HJSE36CH = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-HJSE36CH.js"() {
	init_chunk_FZHIMCK6();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WFMWIRTS.js
var init_chunk_WFMWIRTS = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-WFMWIRTS.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UHDYHGOF.js
var init_chunk_UHDYHGOF = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/chunk-UHDYHGOF.js"() {
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/index.js
var init_dist = __esm({ "../../node_modules/.pnpm/remeda@2.21.2/node_modules/remeda/dist/index.js"() {
	init_chunk_7ZI6JRPB();
	init_chunk_OXJMERKM();
	init_chunk_BSLJB6JE();
	init_chunk_NJXNQM3G();
	init_chunk_6RKHJ2CP();
	init_chunk_QDGUNRDA();
	init_chunk_MYLLMFC7();
	init_chunk_N4JUOEMS();
	init_chunk_YRJ25UV2();
	init_chunk_Q5ASJ5N7();
	init_chunk_WZOX4VKU();
	init_chunk_Y3VKZ3P5();
	init_chunk_ZRKG4NSC();
	init_chunk_QJLMYOTX();
	init_chunk_DM52TTEP();
	init_chunk_2P44HXVH();
	init_chunk_ZJS5DNQW();
	init_chunk_AHDTEY6L();
	init_chunk_4UEQNEAO();
	init_chunk_T4PLMLCP();
	init_chunk_SSDL7ATG();
	init_chunk_EVIH3PFY();
	init_chunk_RBODUO3Q();
	init_chunk_S52RID4A();
	init_chunk_WWPMIW33();
	init_chunk_57KROWWS();
	init_chunk_V6HCOU6T();
	init_chunk_ALS6JP7S();
	init_chunk_QTQTP2VB();
	init_chunk_NFFV4IQT();
	init_chunk_FDH4IRIM();
	init_chunk_YDIA5YQI();
	init_chunk_6OEKBHIX();
	init_chunk_GDGEDZJG();
	init_chunk_XE3XIKTJ();
	init_chunk_HVJXDSOP();
	init_chunk_MSOX5OUI();
	init_chunk_C4OZY4Z2();
	init_chunk_P2PQB7KO();
	init_chunk_UHZ33J57();
	init_chunk_6RL33UFT();
	init_chunk_YNNF733L();
	init_chunk_4YLWJIJ6();
	init_chunk_KVHF7QRD();
	init_chunk_UA6DVSZ3();
	init_chunk_G5B2IDWB();
	init_chunk_3FKP6OOU();
	init_chunk_VFSOOVKJ();
	init_chunk_K3UJMX27();
	init_chunk_LE6I3KC6();
	init_chunk_ENS7GPLZ();
	init_chunk_DSLWSGID();
	init_chunk_R72GEKLP();
	init_chunk_3IFJP4R5();
	init_chunk_J4EKWFDW();
	init_chunk_PGMPBC5Q();
	init_chunk_GGYFZTDW();
	init_chunk_EDOGCRPU();
	init_chunk_4NRWDO7P();
	init_chunk_W6ZHPGFP();
	init_chunk_J3IRE4DI();
	init_chunk_RZUYD7QY();
	init_chunk_A3PVMI4K();
	init_chunk_4ZFFLFWV();
	init_chunk_C6IMN7SF();
	init_chunk_NS6ZBRLP();
	init_chunk_FMPZ2CLX();
	init_chunk_3ZJAREUD();
	init_chunk_ZXVA7VDE();
	init_chunk_KQRZQWDE();
	init_chunk_567G5ZXL();
	init_chunk_3D3RWAVJ();
	init_chunk_K2FFNW24();
	init_chunk_5S4PYKVY();
	init_chunk_5WKPQX7L();
	init_chunk_PDQFB3TV();
	init_chunk_XMLUDZIW();
	init_chunk_GMMLSO2N();
	init_chunk_5NQBDF4H();
	init_chunk_PFSVCZNE();
	init_chunk_VO5MRBXA();
	init_chunk_XUX3ZEXI();
	init_chunk_KZIKCQ56();
	init_chunk_P3DXEVTH();
	init_chunk_6GTAPB47();
	init_chunk_NMC53JVB();
	init_chunk_BZNENX2T();
	init_chunk_PULGOXDA();
	init_chunk_OLNQBNAJ();
	init_chunk_QOEIYQAG();
	init_chunk_SFZGYJFI();
	init_chunk_JJZ7E4YG();
	init_chunk_VCYTMP4D();
	init_chunk_CAZXBO45();
	init_chunk_ENOHV5LT();
	init_chunk_U753ZCO5();
	init_chunk_5DU4ITSF();
	init_chunk_GK5I7C4J();
	init_chunk_HV3WACXG();
	init_chunk_JK3VNB42();
	init_chunk_T45O7BFY();
	init_chunk_OP5ZF26D();
	init_chunk_BO3LQZNF();
	init_chunk_I3D2BSWJ();
	init_chunk_7QX4DO53();
	init_chunk_VMV5GVZ5();
	init_chunk_OWH4IQQW();
	init_chunk_ETADWPSK();
	init_chunk_OFZGPGDZ();
	init_chunk_R7PILVSQ();
	init_chunk_HGKLN5KY();
	init_chunk_HJSE3ESO();
	init_chunk_B6PG574O();
	init_chunk_ICBBHOCR();
	init_chunk_DAE7EJQ2();
	init_chunk_XPCYQPKH();
	init_chunk_FRNNS7AX();
	init_chunk_QJOWZFYO();
	init_chunk_VIBSXWWU();
	init_chunk_T4H4IOYC();
	init_chunk_GPLTWAVR();
	init_chunk_PIX5OHMW();
	init_chunk_VVM5DH6Z();
	init_chunk_PVYOMZ3I();
	init_chunk_P6LAFGAN();
	init_chunk_MQDP6CFS();
	init_chunk_UZ6BOIAH();
	init_chunk_KI5UAETW();
	init_chunk_GYH2VCL4();
	init_chunk_26ILFTOP();
	init_chunk_2KIKGHAO();
	init_chunk_YVMG2XEU();
	init_chunk_ZP6YI2JT();
	init_chunk_OCJMJQKO();
	init_chunk_P5WDBFN2();
	init_chunk_OAMXQXGR();
	init_chunk_J7R2OSHS();
	init_chunk_GIKF2ZNG();
	init_chunk_XWBKJZIP();
	init_chunk_XHPQVWZM();
	init_chunk_2OQBQB3V();
	init_chunk_H4OTHZJB();
	init_chunk_XZ6COQKM();
	init_chunk_T5XG33UI();
	init_chunk_ZTNFU7RH();
	init_chunk_OIQJEOF7();
	init_chunk_GKXRNLHM();
	init_chunk_NYIWN625();
	init_chunk_LFJW7BOT();
	init_chunk_3GOCSNFN();
	init_chunk_WPTI67A4();
	init_chunk_W2ARC73P();
	init_chunk_3UBK2BVM();
	init_chunk_VFECZ57D();
	init_chunk_VG2NVNXT();
	init_chunk_HJSE36CH();
	init_chunk_WFMWIRTS();
	init_chunk_UHDYHGOF();
	init_chunk_WIMGWYZL();
} });

//#endregion
//#region src/plugin/bindingify-hook-filter.ts
function generalHookFilterMatcherToFilterExprs(matcher, stringKind) {
	if (typeof matcher === "string" || matcher instanceof RegExp) return [include(id(matcher))];
	if (Array.isArray(matcher)) return matcher.map((m$10) => include(id(m$10)));
	if (matcher.custom) return matcher.custom;
	let ret = [];
	let isCode = stringKind === "code";
	if (matcher.exclude) ret.push(...arraify(matcher.exclude).map((m$10) => exclude(isCode ? code(m$10) : id(m$10))));
	if (matcher.include) ret.push(...arraify(matcher.include).map((m$10) => include(isCode ? code(m$10) : id(m$10))));
	return ret;
}
function transformFilterMatcherToFilterExprs(filterOption) {
	if (!filterOption) return void 0;
	const { id: id$1, code: code$1, moduleType: moduleType$1, custom } = filterOption;
	if (custom) return custom;
	let ret = [];
	let idIncludes = [];
	let idExcludes = [];
	let codeIncludes = [];
	let codeExcludes = [];
	if (id$1) [idIncludes, idExcludes] = d$3(generalHookFilterMatcherToFilterExprs(id$1, "id") ?? [], (m$10) => m$10.kind === "include");
	if (code$1) [codeIncludes, codeExcludes] = d$3(generalHookFilterMatcherToFilterExprs(code$1, "code") ?? [], (m$10) => m$10.kind === "include");
	ret.push(...idExcludes);
	ret.push(...codeExcludes);
	let cursor;
	if (moduleType$1) {
		let moduleTypes = Array.isArray(moduleType$1) ? moduleType$1 : moduleType$1.include ?? [];
		cursor = joinFilterExprsWithOr(moduleTypes.map((m$10) => moduleType(m$10)));
	}
	if (idIncludes.length) {
		let joinedOrExpr = joinFilterExprsWithOr(idIncludes.map((item) => item.expr));
		if (!cursor) cursor = joinedOrExpr;
		else cursor = and(cursor, joinedOrExpr);
	}
	if (codeIncludes.length) {
		let joinedOrExpr = joinFilterExprsWithOr(codeIncludes.map((item) => item.expr));
		if (!cursor) cursor = joinedOrExpr;
		else cursor = and(cursor, joinedOrExpr);
	}
	if (cursor) ret.push(include(cursor));
	return ret;
}
function joinFilterExprsWithOr(filterExprs) {
	if (filterExprs.length === 1) return filterExprs[0];
	return or(filterExprs[0], joinFilterExprsWithOr(filterExprs.slice(1)));
}
function bindingifyGeneralHookFilter(matcher, stringKind) {
	let filterExprs = generalHookFilterMatcherToFilterExprs(matcher, stringKind);
	let custom = [];
	if (filterExprs) custom = filterExprs.map(bindingifyFilterExpr);
	return { custom: custom.length > 0 ? custom : void 0 };
}
function bindingifyFilterExpr(expr) {
	let list = [];
	bindingifyFilterExprImpl(expr, list);
	return list;
}
function bindingifyFilterExprImpl(expr, list) {
	switch (expr.kind) {
		case "and": {
			bindingifyFilterExprImpl(expr.right, list);
			bindingifyFilterExprImpl(expr.left, list);
			list.push({ kind: "And" });
			break;
		}
		case "not": {
			bindingifyFilterExprImpl(expr.expr, list);
			list.push({ kind: "Not" });
			break;
		}
		case "id": {
			list.push({
				kind: "Id",
				value: expr.pattern
			});
			break;
		}
		case "moduleType": {
			list.push({
				kind: "ModuleType",
				value: expr.pattern
			});
			break;
		}
		case "code": {
			list.push({
				kind: "Code",
				value: expr.pattern
			});
			break;
		}
		case "include": {
			bindingifyFilterExprImpl(expr.expr, list);
			list.push({ kind: "Include" });
			break;
		}
		case "exclude": {
			bindingifyFilterExprImpl(expr.expr, list);
			list.push({ kind: "Exclude" });
			break;
		}
		default: throw new Error(`Unknown filter expression kind: ${expr.kind}`);
	}
}
function bindingifyResolveIdFilter(filterOption) {
	return filterOption?.id ? bindingifyGeneralHookFilter(filterOption.id, "id") : void 0;
}
function bindingifyLoadFilter(filterOption) {
	return filterOption?.id ? bindingifyGeneralHookFilter(filterOption.id, "id") : void 0;
}
function bindingifyTransformFilter(filterOption) {
	if (!filterOption) return void 0;
	let custom = transformFilterMatcherToFilterExprs(filterOption);
	let ret = [];
	if (custom) ret = custom.map(bindingifyFilterExpr);
	return { custom: ret.length > 0 ? ret : void 0 };
}
function bindingifyRenderChunkFilter(filterOption) {
	if (filterOption) {
		const { code: code$1 } = filterOption;
		return { custom: code$1 ? bindingifyGeneralHookFilter(code$1, "code").custom : void 0 };
	}
}
var init_bindingify_hook_filter = __esm({ "src/plugin/bindingify-hook-filter.ts"() {
	init_dist();
	init_filter_expression_index();
	init_misc();
} });

//#endregion
//#region src/plugin/bindingify-plugin-hook-meta.ts
function bindingifyPluginHookMeta(options) {
	return { order: bindingPluginOrder(options.order) };
}
function bindingPluginOrder(order) {
	switch (order) {
		case "post": return import_binding.BindingPluginOrder.Post;
		case "pre": return import_binding.BindingPluginOrder.Pre;
		case null:
		case void 0: return void 0;
		default: throw new Error(`Unknown plugin order: ${order}`);
	}
}
var init_bindingify_plugin_hook_meta = __esm({ "src/plugin/bindingify-plugin-hook-meta.ts"() {} });

//#endregion
//#region src/utils/asset-source.ts
function transformAssetSource(bindingAssetSource$1) {
	return bindingAssetSource$1.inner;
}
function bindingAssetSource(source) {
	return { inner: source };
}
var init_asset_source = __esm({ "src/utils/asset-source.ts"() {} });

//#endregion
//#region src/plugin/plugin-context.ts
var PluginContextImpl;
var init_plugin_context = __esm({ "src/plugin/plugin-context.ts"() {
	init_logging();
	init_logs();
	init_parse_ast_index();
	init_minimal_plugin_context();
	init_asset_source();
	init_misc();
	init_transform_side_effects();
	PluginContextImpl = class extends MinimalPluginContextImpl {
		getModuleInfo;
		constructor(outputOptions, context, plugin, data, onLog, logLevel, watchMode, currentLoadingModule) {
			super(onLog, logLevel, plugin.name, watchMode);
			this.outputOptions = outputOptions;
			this.context = context;
			this.data = data;
			this.onLog = onLog;
			this.currentLoadingModule = currentLoadingModule;
			this.getModuleInfo = (id$1) => this.data.getModuleInfo(id$1, context);
		}
		async load(options) {
			const id$1 = options.id;
			if (id$1 === this.currentLoadingModule) this.onLog(LOG_LEVEL_WARN, logCycleLoading(this.pluginName, this.currentLoadingModule));
			const moduleInfo = this.data.getModuleInfo(id$1, this.context);
			if (moduleInfo && moduleInfo.code !== null) return moduleInfo;
			const rawOptions = {
				meta: options.meta || {},
				moduleSideEffects: options.moduleSideEffects || null,
				invalidate: false
			};
			this.data.updateModuleOption(id$1, rawOptions);
			async function createLoadModulePromise(context, data) {
				const loadPromise = data.loadModulePromiseMap.get(id$1);
				if (loadPromise) return loadPromise;
				const promise = new Promise((resolve, _) => {
					data.loadModulePromiseResolveFnMap.set(id$1, resolve);
				});
				data.loadModulePromiseMap.set(id$1, promise);
				try {
					await context.load(id$1, bindingifySideEffects(options.moduleSideEffects));
				} catch (e$5) {
					data.loadModulePromiseMap.delete(id$1);
					data.loadModulePromiseResolveFnMap.delete(id$1);
					throw e$5;
				}
				return promise;
			}
			await createLoadModulePromise(this.context, this.data);
			return this.data.getModuleInfo(id$1, this.context);
		}
		async resolve(source, importer, options) {
			let receipt = void 0;
			if (options != null) receipt = this.data.saveResolveOptions(options);
			const res = await this.context.resolve(source, importer, {
				custom: receipt,
				skipSelf: options?.skipSelf
			});
			if (receipt != null) this.data.removeSavedResolveOptions(receipt);
			if (res == null) return null;
			const info = this.data.getModuleOption(res.id) || {};
			return {
				...res,
				external: res.external === "relative" ? unreachable(`The PluginContext resolve result external couldn't be 'relative'`) : res.external,
				...info
			};
		}
		emitFile = (file) => {
			if (file.type === "prebuilt-chunk") return unimplemented("PluginContext.emitFile with type prebuilt-chunk");
			if (file.type === "chunk") return this.context.emitChunk(file);
			const fnSanitizedFileName = file.fileName || typeof this.outputOptions.sanitizeFileName !== "function" ? void 0 : this.outputOptions.sanitizeFileName(file.name || "asset");
			const filename = file.fileName ? void 0 : this.getAssetFileNames(file);
			return this.context.emitFile({
				...file,
				originalFileName: file.originalFileName || void 0,
				source: bindingAssetSource(file.source)
			}, filename, fnSanitizedFileName);
		};
		getAssetFileNames(file) {
			if (typeof this.outputOptions.assetFileNames === "function") return this.outputOptions.assetFileNames({
				names: file.name ? [file.name] : [],
				originalFileNames: file.originalFileName ? [file.originalFileName] : [],
				source: file.source,
				type: "asset"
			});
		}
		getFileName(referenceId) {
			return this.context.getFileName(referenceId);
		}
		getModuleIds() {
			return this.data.getModuleIds(this.context);
		}
		addWatchFile(id$1) {
			this.context.addWatchFile(id$1);
		}
		parse(input, options) {
			return parseAst(input, options);
		}
	};
} });

//#endregion
//#region src/plugin/transform-plugin-context.ts
var TransformPluginContextImpl;
var init_transform_plugin_context = __esm({ "src/plugin/transform-plugin-context.ts"() {
	init_log_handler();
	init_logs();
	init_plugin_context();
	TransformPluginContextImpl = class extends PluginContextImpl {
		constructor(outputOptions, context, plugin, data, inner, moduleId, moduleSource, onLog, LogLevelOption, watchMode) {
			super(outputOptions, context, plugin, data, onLog, LogLevelOption, watchMode, moduleId);
			this.inner = inner;
			this.moduleId = moduleId;
			this.moduleSource = moduleSource;
			const getLogHandler$1 = (handler) => (log, pos) => {
				log = normalizeLog(log);
				if (pos) augmentCodeLocation(log, pos, moduleSource, moduleId);
				log.id = moduleId;
				log.hook = "transform";
				handler(log);
			};
			this.debug = getLogHandler$1(this.debug);
			this.warn = getLogHandler$1(this.warn);
			this.info = getLogHandler$1(this.info);
		}
		error(e$5, pos) {
			if (typeof e$5 === "string") e$5 = { message: e$5 };
			if (pos) augmentCodeLocation(e$5, pos, this.moduleSource, this.moduleId);
			e$5.id = this.moduleId;
			e$5.hook = "transform";
			return error(logPluginError(normalizeLog(e$5), this.pluginName));
		}
		getCombinedSourcemap() {
			return JSON.parse(this.inner.getCombinedSourcemap());
		}
	};
} });

//#endregion
//#region src/plugin/bindingify-build-hooks.ts
function bindingifyBuildStart(args) {
	const hook = args.plugin.buildStart;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, opts) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), new NormalizedInputOptionsImpl(opts, args.onLog));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyBuildEnd(args) {
	const hook = args.plugin.buildEnd;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, err) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), err ? normalizeErrors(err) : void 0);
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyResolveId(args) {
	const hook = args.plugin.resolveId;
	if (!hook) return {};
	const { handler, meta, options } = normalizeHook(hook);
	return {
		plugin: async (ctx, specifier, importer, extraOptions) => {
			const contextResolveOptions = extraOptions.custom != null ? args.pluginContextData.getSavedResolveOptions(extraOptions.custom) : void 0;
			const newExtraOptions = {
				...extraOptions,
				custom: contextResolveOptions?.custom,
				[SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF]: contextResolveOptions?.[SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF]
			};
			const ret = await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), specifier, importer ?? void 0, newExtraOptions);
			if (ret == null) return;
			if (ret === false) return {
				id: specifier,
				external: true,
				normalizeExternalId: true
			};
			if (typeof ret === "string") return {
				id: ret,
				normalizeExternalId: true
			};
			let exist = args.pluginContextData.updateModuleOption(ret.id, {
				meta: ret.meta || {},
				moduleSideEffects: ret.moduleSideEffects ?? null,
				invalidate: false
			});
			return {
				id: ret.id,
				external: ret.external,
				normalizeExternalId: false,
				sideEffects: bindingifySideEffects(exist.moduleSideEffects)
			};
		},
		meta: bindingifyPluginHookMeta(meta),
		filter: bindingifyResolveIdFilter(options.filter)
	};
}
function bindingifyResolveDynamicImport(args) {
	const hook = args.plugin.resolveDynamicImport;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, specifier, importer) => {
			const ret = await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), specifier, importer ?? void 0);
			if (ret == null) return;
			if (ret === false) return {
				id: specifier,
				external: true
			};
			if (typeof ret === "string") return { id: ret };
			const result = {
				id: ret.id,
				external: ret.external
			};
			if (ret.moduleSideEffects !== null) result.sideEffects = bindingifySideEffects(ret.moduleSideEffects);
			args.pluginContextData.updateModuleOption(ret.id, {
				meta: ret.meta || {},
				moduleSideEffects: ret.moduleSideEffects || null,
				invalidate: false
			});
			return result;
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyTransform(args) {
	const hook = args.plugin.transform;
	if (!hook) return {};
	const { handler, meta, options } = normalizeHook(hook);
	return {
		plugin: async (ctx, code$1, id$1, meta$1) => {
			const ret = await handler.call(new TransformPluginContextImpl(args.outputOptions, ctx.inner(), args.plugin, args.pluginContextData, ctx, id$1, code$1, args.onLog, args.logLevel, args.watchMode), code$1, id$1, meta$1);
			if (ret == null) return void 0;
			if (typeof ret === "string") return { code: ret };
			let moduleOption = args.pluginContextData.updateModuleOption(id$1, {
				meta: ret.meta ?? {},
				moduleSideEffects: ret.moduleSideEffects ?? null,
				invalidate: false
			});
			return {
				code: ret.code,
				map: bindingifySourcemap(normalizeTransformHookSourcemap(id$1, code$1, ret.map)),
				sideEffects: bindingifySideEffects(moduleOption.moduleSideEffects),
				moduleType: ret.moduleType
			};
		},
		meta: bindingifyPluginHookMeta(meta),
		filter: bindingifyTransformFilter(options.filter)
	};
}
function bindingifyLoad(args) {
	const hook = args.plugin.load;
	if (!hook) return {};
	const { handler, meta, options } = normalizeHook(hook);
	return {
		plugin: async (ctx, id$1) => {
			const ret = await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode, id$1), id$1);
			if (ret == null) return;
			if (typeof ret === "string") return { code: ret };
			let moduleOption = args.pluginContextData.updateModuleOption(id$1, {
				meta: ret.meta || {},
				moduleSideEffects: ret.moduleSideEffects ?? null,
				invalidate: false
			});
			let map = preProcessSourceMap(ret, id$1);
			return {
				code: ret.code,
				map: bindingifySourcemap(map),
				moduleType: ret.moduleType,
				sideEffects: bindingifySideEffects(moduleOption.moduleSideEffects)
			};
		},
		meta: bindingifyPluginHookMeta(meta),
		filter: bindingifyLoadFilter(options.filter)
	};
}
function preProcessSourceMap(ret, id$1) {
	if (!ret.map) return;
	let map = typeof ret.map === "object" ? ret.map : JSON.parse(ret.map);
	if (!isEmptySourcemapFiled(map.sources)) {
		const directory = path.dirname(id$1) || ".";
		const sourceRoot = map.sourceRoot || ".";
		map.sources = map.sources.map((source) => path.resolve(directory, sourceRoot, source));
	}
	return map;
}
function bindingifyModuleParsed(args) {
	const hook = args.plugin.moduleParsed;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, moduleInfo) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformModuleInfo(moduleInfo, args.pluginContextData.getModuleOption(moduleInfo.id)));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
var init_bindingify_build_hooks = __esm({ "src/plugin/bindingify-build-hooks.ts"() {
	init_normalize_hook();
	init_plugin_context$1();
	init_normalized_input_options();
	init_sourcemap();
	init_error();
	init_transform_module_info();
	init_transform_side_effects();
	init_transform_sourcemap();
	init_bindingify_hook_filter();
	init_bindingify_plugin_hook_meta();
	init_plugin_context();
	init_transform_plugin_context();
} });

//#endregion
//#region src/utils/transform-rendered-module.ts
function transformToRenderedModule(bindingRenderedModule) {
	return {
		get code() {
			return bindingRenderedModule.code;
		},
		get renderedLength() {
			return bindingRenderedModule.code?.length || 0;
		},
		get renderedExports() {
			return bindingRenderedModule.renderedExports;
		}
	};
}
var init_transform_rendered_module = __esm({ "src/utils/transform-rendered-module.ts"() {} });

//#endregion
//#region src/utils/transform-rendered-chunk.ts
function transformRenderedChunk(chunk) {
	let modules = null;
	return {
		get name() {
			return chunk.name;
		},
		get isEntry() {
			return chunk.isEntry;
		},
		get isDynamicEntry() {
			return chunk.isDynamicEntry;
		},
		get facadeModuleId() {
			return chunk.facadeModuleId;
		},
		get moduleIds() {
			return chunk.moduleIds;
		},
		get exports() {
			return chunk.exports;
		},
		get fileName() {
			return chunk.fileName;
		},
		get imports() {
			return chunk.imports;
		},
		get dynamicImports() {
			return chunk.dynamicImports;
		},
		get modules() {
			if (!modules) modules = transformChunkModules(chunk.modules);
			return modules;
		}
	};
}
function transformChunkModules(modules) {
	const result = {};
	for (let i$22 = 0; i$22 < modules.values.length; i$22++) {
		let key = modules.keys[i$22];
		const mod = modules.values[i$22];
		result[key] = transformToRenderedModule(mod);
	}
	return result;
}
var init_transform_rendered_chunk = __esm({ "src/utils/transform-rendered-chunk.ts"() {
	init_transform_rendered_module();
} });

//#endregion
//#region src/utils/bindingify-output-options.ts
function bindingifyOutputOptions(outputOptions) {
	const { dir, format, exports, hashCharacters, sourcemap, sourcemapDebugIds, sourcemapIgnoreList, sourcemapPathTransform, name, assetFileNames, entryFileNames, chunkFileNames, cssEntryFileNames, cssChunkFileNames, banner, footer, intro, outro, esModule, globals, file, sanitizeFileName } = outputOptions;
	return {
		dir,
		file: file == null ? void 0 : file,
		format: bindingifyFormat(format),
		exports,
		hashCharacters,
		sourcemap: bindingifySourcemap$1(sourcemap),
		sourcemapDebugIds,
		sourcemapIgnoreList: bindingifySourcemapIgnoreList(sourcemapIgnoreList),
		sourcemapPathTransform,
		banner: bindingifyAddon(banner),
		footer: bindingifyAddon(footer),
		intro: bindingifyAddon(intro),
		outro: bindingifyAddon(outro),
		extend: outputOptions.extend,
		globals,
		esModule,
		name,
		assetFileNames: bindingifyAssetFilenames(assetFileNames),
		entryFileNames,
		chunkFileNames,
		cssEntryFileNames,
		cssChunkFileNames,
		plugins: [],
		minify: outputOptions.minify,
		externalLiveBindings: outputOptions.externalLiveBindings,
		inlineDynamicImports: outputOptions.inlineDynamicImports,
		advancedChunks: outputOptions.advancedChunks,
		polyfillRequire: outputOptions.polyfillRequire,
		target: outputOptions.target,
		sanitizeFileName
	};
}
function bindingifyAddon(configAddon) {
	return async (chunk) => {
		if (typeof configAddon === "function") return configAddon(transformRenderedChunk(chunk));
		return configAddon || "";
	};
}
function bindingifyFormat(format) {
	switch (format) {
		case void 0:
		case "es":
		case "esm":
		case "module": return "es";
		case "cjs":
		case "commonjs": return "cjs";
		case "iife": return "iife";
		case "umd": return "umd";
		case "experimental-app": return "app";
		default: unimplemented(`output.format: ${format}`);
	}
}
function bindingifySourcemap$1(sourcemap) {
	switch (sourcemap) {
		case true: return "file";
		case "inline": return "inline";
		case false:
		case void 0: return void 0;
		case "hidden": return "hidden";
		default: throw new Error(`unknown sourcemap: ${sourcemap}`);
	}
}
function bindingifySourcemapIgnoreList(sourcemapIgnoreList) {
	return typeof sourcemapIgnoreList === "function" ? sourcemapIgnoreList : sourcemapIgnoreList === false ? () => false : (relativeSourcePath, _sourcemapPath) => relativeSourcePath.includes("node_modules");
}
function bindingifyAssetFilenames(assetFileNames) {
	if (typeof assetFileNames === "function") return (asset) => {
		return assetFileNames({
			names: asset.names,
			originalFileNames: asset.originalFileNames,
			source: transformAssetSource(asset.source),
			type: "asset"
		});
	};
	return assetFileNames;
}
var init_bindingify_output_options = __esm({ "src/utils/bindingify-output-options.ts"() {
	init_asset_source();
	init_misc();
	init_transform_rendered_chunk();
} });

//#endregion
//#region src/options/normalized-output-options.ts
function normalizeAddon(value) {
	if (typeof value === "function") return value;
	return () => value || "";
}
var NormalizedOutputOptionsImpl;
var init_normalized_output_options = __esm({ "src/options/normalized-output-options.ts"() {
	init_bindingify_output_options();
	NormalizedOutputOptionsImpl = class {
		constructor(inner, outputOptions, normalizedOutputPlugins) {
			this.inner = inner;
			this.outputOptions = outputOptions;
			this.normalizedOutputPlugins = normalizedOutputPlugins;
		}
		get dir() {
			return this.inner.dir ?? void 0;
		}
		get entryFileNames() {
			return this.inner.entryFilenames || this.outputOptions.entryFileNames;
		}
		get chunkFileNames() {
			return this.inner.chunkFilenames || this.outputOptions.chunkFileNames;
		}
		get assetFileNames() {
			return this.inner.assetFilenames || this.outputOptions.assetFileNames;
		}
		get format() {
			return this.inner.format;
		}
		get exports() {
			return this.inner.exports;
		}
		get sourcemap() {
			return this.inner.sourcemap;
		}
		get cssEntryFileNames() {
			return this.inner.cssEntryFilenames || this.outputOptions.cssEntryFileNames;
		}
		get cssChunkFileNames() {
			return this.inner.cssChunkFilenames || this.outputOptions.cssChunkFileNames;
		}
		get shimMissingExports() {
			return this.inner.shimMissingExports;
		}
		get name() {
			return this.inner.name ?? void 0;
		}
		get file() {
			return this.inner.file ?? void 0;
		}
		get inlineDynamicImports() {
			return this.inner.inlineDynamicImports;
		}
		get externalLiveBindings() {
			return this.inner.externalLiveBindings;
		}
		get banner() {
			return normalizeAddon(this.outputOptions.banner);
		}
		get footer() {
			return normalizeAddon(this.outputOptions.footer);
		}
		get intro() {
			return normalizeAddon(this.outputOptions.intro);
		}
		get outro() {
			return normalizeAddon(this.outputOptions.outro);
		}
		get esModule() {
			return this.inner.esModule;
		}
		get extend() {
			return this.inner.extend;
		}
		get globals() {
			return this.inner.globals || this.outputOptions.globals;
		}
		get hashCharacters() {
			return this.inner.hashCharacters;
		}
		get sourcemapDebugIds() {
			return this.inner.sourcemapDebugIds;
		}
		get sourcemapIgnoreList() {
			return bindingifySourcemapIgnoreList(this.outputOptions.sourcemapIgnoreList);
		}
		get sourcemapPathTransform() {
			return this.outputOptions.sourcemapPathTransform;
		}
		get minify() {
			return this.inner.minify;
		}
		get comments() {
			return this.inner.comments;
		}
		get polyfillRequire() {
			return this.inner.polyfillRequire;
		}
		get plugins() {
			return this.normalizedOutputPlugins;
		}
	};
} });

//#endregion
//#region src/utils/transform-to-rollup-output.ts
function transformToRollupSourceMap(map) {
	const parsed = JSON.parse(map);
	const obj = {
		...parsed,
		toString() {
			return JSON.stringify(obj);
		},
		toUrl() {
			return `data:application/json;charset=utf-8;base64,${Buffer.from(obj.toString(), "utf-8").toString("base64")}`;
		}
	};
	return obj;
}
function transformToRollupOutputChunk(bindingChunk, changed) {
	const chunk = {
		type: "chunk",
		get code() {
			return bindingChunk.code;
		},
		fileName: bindingChunk.fileName,
		name: bindingChunk.name,
		get modules() {
			return transformChunkModules(bindingChunk.modules);
		},
		get imports() {
			return bindingChunk.imports;
		},
		get dynamicImports() {
			return bindingChunk.dynamicImports;
		},
		exports: bindingChunk.exports,
		isEntry: bindingChunk.isEntry,
		facadeModuleId: bindingChunk.facadeModuleId || null,
		isDynamicEntry: bindingChunk.isDynamicEntry,
		get moduleIds() {
			return bindingChunk.moduleIds;
		},
		get map() {
			return bindingChunk.map ? transformToRollupSourceMap(bindingChunk.map) : null;
		},
		sourcemapFileName: bindingChunk.sourcemapFileName || null,
		preliminaryFileName: bindingChunk.preliminaryFileName
	};
	const cache = {};
	return new Proxy(chunk, {
		get(target, p$3) {
			if (p$3 in cache) return cache[p$3];
			const value = target[p$3];
			cache[p$3] = value;
			return value;
		},
		set(target, p$3, newValue) {
			cache[p$3] = newValue;
			changed?.updated.add(bindingChunk.fileName);
			return true;
		},
		has(target, p$3) {
			if (p$3 in cache) return true;
			return p$3 in target;
		}
	});
}
function transformToRollupOutputAsset(bindingAsset, changed) {
	const asset = {
		type: "asset",
		fileName: bindingAsset.fileName,
		originalFileName: bindingAsset.originalFileName || null,
		originalFileNames: bindingAsset.originalFileNames,
		get source() {
			return transformAssetSource(bindingAsset.source);
		},
		name: bindingAsset.name ?? void 0,
		names: bindingAsset.names
	};
	const cache = {};
	return new Proxy(asset, {
		get(target, p$3) {
			if (p$3 in cache) return cache[p$3];
			const value = target[p$3];
			cache[p$3] = value;
			return value;
		},
		set(target, p$3, newValue) {
			cache[p$3] = newValue;
			changed?.updated.add(bindingAsset.fileName);
			return true;
		}
	});
}
function transformToRollupOutput(output, changed) {
	handleOutputErrors(output);
	const { chunks, assets } = output;
	return { output: [...chunks.map((chunk) => transformToRollupOutputChunk(chunk, changed)), ...assets.map((asset) => transformToRollupOutputAsset(asset, changed))] };
}
function handleOutputErrors(output) {
	const rawErrors = output.errors;
	if (rawErrors.length > 0) throw normalizeErrors(rawErrors);
}
function transformToOutputBundle(output, changed) {
	const bundle = Object.fromEntries(transformToRollupOutput(output, changed).output.map((item) => [item.fileName, item]));
	return new Proxy(bundle, { deleteProperty(target, property) {
		if (typeof property === "string") changed.deleted.add(property);
		return true;
	} });
}
function collectChangedBundle(changed, bundle) {
	const assets = [];
	const chunks = [];
	for (const key in bundle) {
		if (changed.deleted.has(key) || !changed.updated.has(key)) continue;
		const item = bundle[key];
		if (item.type === "asset") assets.push({
			filename: item.fileName,
			originalFileNames: item.originalFileNames,
			source: bindingAssetSource(item.source),
			names: item.names
		});
		else chunks.push({
			code: item.code,
			filename: item.fileName,
			name: item.name,
			isEntry: item.isEntry,
			exports: item.exports,
			modules: {},
			imports: item.imports,
			dynamicImports: item.dynamicImports,
			facadeModuleId: item.facadeModuleId || void 0,
			isDynamicEntry: item.isDynamicEntry,
			moduleIds: item.moduleIds,
			map: bindingifySourcemap(item.map),
			sourcemapFilename: item.sourcemapFileName || void 0,
			preliminaryFilename: item.preliminaryFileName
		});
	}
	return {
		assets,
		chunks,
		deleted: Array.from(changed.deleted)
	};
}
var init_transform_to_rollup_output = __esm({ "src/utils/transform-to-rollup-output.ts"() {
	init_sourcemap();
	init_asset_source();
	init_error();
	init_transform_rendered_chunk();
} });

//#endregion
//#region src/plugin/bindingify-output-hooks.ts
function bindingifyRenderStart(args) {
	const hook = args.plugin.renderStart;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, opts) => {
			handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), new NormalizedOutputOptionsImpl(opts, args.outputOptions, args.normalizedOutputPlugins), new NormalizedInputOptionsImpl(opts, args.onLog));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyRenderChunk(args) {
	const hook = args.plugin.renderChunk;
	if (!hook) return {};
	const { handler, meta, options } = normalizeHook(hook);
	return {
		plugin: async (ctx, code$1, chunk, opts, meta$1) => {
			if (args.pluginContextData.getRenderChunkMeta() == null) args.pluginContextData.setRenderChunkMeta({ chunks: Object.fromEntries(Object.entries(meta$1.chunks).map(([key, value]) => [key, transformRenderedChunk(value)])) });
			const ret = await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), code$1, transformRenderedChunk(chunk), new NormalizedOutputOptionsImpl(opts, args.outputOptions, args.normalizedOutputPlugins), args.pluginContextData.getRenderChunkMeta());
			if (ret == null) return;
			if (typeof ret === "string") return { code: ret };
			if (!ret.map) return { code: ret.code };
			return {
				code: ret.code,
				map: bindingifySourcemap(ret.map)
			};
		},
		meta: bindingifyPluginHookMeta(meta),
		filter: bindingifyRenderChunkFilter(options.filter)
	};
}
function bindingifyAugmentChunkHash(args) {
	const hook = args.plugin.augmentChunkHash;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, chunk) => {
			return await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformRenderedChunk(chunk));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyRenderError(args) {
	const hook = args.plugin.renderError;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, err) => {
			handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), normalizeErrors(err));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyGenerateBundle(args) {
	const hook = args.plugin.generateBundle;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, bundle, isWrite, opts) => {
			const changed = {
				updated: new Set(),
				deleted: new Set()
			};
			const output = transformToOutputBundle(bundle, changed);
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), new NormalizedOutputOptionsImpl(opts, args.outputOptions, args.normalizedOutputPlugins), output, isWrite);
			return collectChangedBundle(changed, output);
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyWriteBundle(args) {
	const hook = args.plugin.writeBundle;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, bundle, opts) => {
			const changed = {
				updated: new Set(),
				deleted: new Set()
			};
			const output = transformToOutputBundle(bundle, changed);
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), new NormalizedOutputOptionsImpl(opts, args.outputOptions, args.normalizedOutputPlugins), output);
			return collectChangedBundle(changed, output);
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyCloseBundle(args) {
	const hook = args.plugin.closeBundle;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyBanner(args) {
	const hook = args.plugin.banner;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, chunk) => {
			if (typeof handler === "string") return handler;
			return handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformRenderedChunk(chunk));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyFooter(args) {
	const hook = args.plugin.footer;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, chunk) => {
			if (typeof handler === "string") return handler;
			return handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformRenderedChunk(chunk));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyIntro(args) {
	const hook = args.plugin.intro;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, chunk) => {
			if (typeof handler === "string") return handler;
			return handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformRenderedChunk(chunk));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyOutro(args) {
	const hook = args.plugin.outro;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, chunk) => {
			if (typeof handler === "string") return handler;
			return handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), transformRenderedChunk(chunk));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
var init_bindingify_output_hooks = __esm({ "src/plugin/bindingify-output-hooks.ts"() {
	init_normalized_input_options();
	init_normalized_output_options();
	init_sourcemap();
	init_error();
	init_normalize_hook();
	init_transform_rendered_chunk();
	init_transform_to_rollup_output();
	init_bindingify_hook_filter();
	init_bindingify_plugin_hook_meta();
	init_plugin_context();
} });

//#endregion
//#region src/plugin/bindingify-watch-hooks.ts
function bindingifyWatchChange(args) {
	const hook = args.plugin.watchChange;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx, id$1, event) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode), id$1, { event });
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
function bindingifyCloseWatcher(args) {
	const hook = args.plugin.closeWatcher;
	if (!hook) return {};
	const { handler, meta } = normalizeHook(hook);
	return {
		plugin: async (ctx) => {
			await handler.call(new PluginContextImpl(args.outputOptions, ctx, args.plugin, args.pluginContextData, args.onLog, args.logLevel, args.watchMode));
		},
		meta: bindingifyPluginHookMeta(meta)
	};
}
var init_bindingify_watch_hooks = __esm({ "src/plugin/bindingify-watch-hooks.ts"() {
	init_normalize_hook();
	init_bindingify_plugin_hook_meta();
	init_plugin_context();
} });

//#endregion
//#region src/plugin/generated/hook-usage.ts
function extractHookUsage(plugin) {
	let hookUsage = new HookUsage();
	if (plugin.buildStart) hookUsage.union(HookUsageKind.buildStart);
	if (plugin.resolveId) hookUsage.union(HookUsageKind.resolveId);
	if (plugin.resolveDynamicImport) hookUsage.union(HookUsageKind.resolveDynamicImport);
	if (plugin.load) hookUsage.union(HookUsageKind.load);
	if (plugin.transform) hookUsage.union(HookUsageKind.transform);
	if (plugin.moduleParsed) hookUsage.union(HookUsageKind.moduleParsed);
	if (plugin.buildEnd) hookUsage.union(HookUsageKind.buildEnd);
	if (plugin.renderStart) hookUsage.union(HookUsageKind.renderStart);
	if (plugin.renderError) hookUsage.union(HookUsageKind.renderError);
	if (plugin.renderChunk) hookUsage.union(HookUsageKind.renderChunk);
	if (plugin.augmentChunkHash) hookUsage.union(HookUsageKind.augmentChunkHash);
	if (plugin.generateBundle) hookUsage.union(HookUsageKind.generateBundle);
	if (plugin.writeBundle) hookUsage.union(HookUsageKind.writeBundle);
	if (plugin.closeBundle) hookUsage.union(HookUsageKind.closeBundle);
	if (plugin.watchChange) hookUsage.union(HookUsageKind.watchChange);
	if (plugin.closeWatcher) hookUsage.union(HookUsageKind.closeWatcher);
	if (plugin.banner) hookUsage.union(HookUsageKind.banner);
	if (plugin.footer) hookUsage.union(HookUsageKind.footer);
	if (plugin.intro) hookUsage.union(HookUsageKind.intro);
	if (plugin.outro) hookUsage.union(HookUsageKind.outro);
	return hookUsage;
}
var HookUsageKind, HookUsage;
var init_hook_usage = __esm({ "src/plugin/generated/hook-usage.ts"() {
	HookUsageKind = /* @__PURE__ */ function(HookUsageKind$1) {
		HookUsageKind$1[HookUsageKind$1["buildStart"] = 1] = "buildStart";
		HookUsageKind$1[HookUsageKind$1["resolveId"] = 2] = "resolveId";
		HookUsageKind$1[HookUsageKind$1["resolveDynamicImport"] = 4] = "resolveDynamicImport";
		HookUsageKind$1[HookUsageKind$1["load"] = 8] = "load";
		HookUsageKind$1[HookUsageKind$1["transform"] = 16] = "transform";
		HookUsageKind$1[HookUsageKind$1["moduleParsed"] = 32] = "moduleParsed";
		HookUsageKind$1[HookUsageKind$1["buildEnd"] = 64] = "buildEnd";
		HookUsageKind$1[HookUsageKind$1["renderStart"] = 128] = "renderStart";
		HookUsageKind$1[HookUsageKind$1["renderError"] = 256] = "renderError";
		HookUsageKind$1[HookUsageKind$1["renderChunk"] = 512] = "renderChunk";
		HookUsageKind$1[HookUsageKind$1["augmentChunkHash"] = 1024] = "augmentChunkHash";
		HookUsageKind$1[HookUsageKind$1["generateBundle"] = 2048] = "generateBundle";
		HookUsageKind$1[HookUsageKind$1["writeBundle"] = 4096] = "writeBundle";
		HookUsageKind$1[HookUsageKind$1["closeBundle"] = 8192] = "closeBundle";
		HookUsageKind$1[HookUsageKind$1["watchChange"] = 16384] = "watchChange";
		HookUsageKind$1[HookUsageKind$1["closeWatcher"] = 32768] = "closeWatcher";
		HookUsageKind$1[HookUsageKind$1["transformAst"] = 65536] = "transformAst";
		HookUsageKind$1[HookUsageKind$1["banner"] = 131072] = "banner";
		HookUsageKind$1[HookUsageKind$1["footer"] = 262144] = "footer";
		HookUsageKind$1[HookUsageKind$1["intro"] = 524288] = "intro";
		HookUsageKind$1[HookUsageKind$1["outro"] = 1048576] = "outro";
		return HookUsageKind$1;
	}({});
	HookUsage = class {
		bitflag = BigInt(0);
		constructor() {}
		union(kind) {
			this.bitflag |= BigInt(kind);
		}
		inner() {
			return Number(this.bitflag);
		}
	};
} });

//#endregion
//#region src/plugin/bindingify-plugin.ts
function bindingifyPlugin(plugin, options, outputOptions, pluginContextData, normalizedOutputPlugins, onLog, logLevel, watchMode) {
	const args = {
		plugin,
		options,
		outputOptions,
		pluginContextData,
		onLog,
		logLevel,
		watchMode,
		normalizedOutputPlugins
	};
	const { plugin: buildStart, meta: buildStartMeta } = bindingifyBuildStart(args);
	const { plugin: resolveId, meta: resolveIdMeta, filter: resolveIdFilter } = bindingifyResolveId(args);
	const { plugin: resolveDynamicImport, meta: resolveDynamicImportMeta } = bindingifyResolveDynamicImport(args);
	const { plugin: buildEnd, meta: buildEndMeta } = bindingifyBuildEnd(args);
	const { plugin: transform, meta: transformMeta, filter: transformFilter } = bindingifyTransform(args);
	const { plugin: moduleParsed, meta: moduleParsedMeta } = bindingifyModuleParsed(args);
	const { plugin: load, meta: loadMeta, filter: loadFilter } = bindingifyLoad(args);
	const { plugin: renderChunk, meta: renderChunkMeta, filter: renderChunkFilter } = bindingifyRenderChunk(args);
	const { plugin: augmentChunkHash, meta: augmentChunkHashMeta } = bindingifyAugmentChunkHash(args);
	const { plugin: renderStart, meta: renderStartMeta } = bindingifyRenderStart(args);
	const { plugin: renderError, meta: renderErrorMeta } = bindingifyRenderError(args);
	const { plugin: generateBundle, meta: generateBundleMeta } = bindingifyGenerateBundle(args);
	const { plugin: writeBundle, meta: writeBundleMeta } = bindingifyWriteBundle(args);
	const { plugin: closeBundle, meta: closeBundleMeta } = bindingifyCloseBundle(args);
	const { plugin: banner, meta: bannerMeta } = bindingifyBanner(args);
	const { plugin: footer, meta: footerMeta } = bindingifyFooter(args);
	const { plugin: intro, meta: introMeta } = bindingifyIntro(args);
	const { plugin: outro, meta: outroMeta } = bindingifyOutro(args);
	const { plugin: watchChange, meta: watchChangeMeta } = bindingifyWatchChange(args);
	const { plugin: closeWatcher, meta: closeWatcherMeta } = bindingifyCloseWatcher(args);
	let hookUsage = extractHookUsage(plugin).inner();
	const result = {
		name: plugin.name,
		buildStart,
		buildStartMeta,
		resolveId,
		resolveIdMeta,
		resolveIdFilter,
		resolveDynamicImport,
		resolveDynamicImportMeta,
		buildEnd,
		buildEndMeta,
		transform,
		transformMeta,
		transformFilter,
		moduleParsed,
		moduleParsedMeta,
		load,
		loadMeta,
		loadFilter,
		renderChunk,
		renderChunkMeta,
		renderChunkFilter,
		augmentChunkHash,
		augmentChunkHashMeta,
		renderStart,
		renderStartMeta,
		renderError,
		renderErrorMeta,
		generateBundle,
		generateBundleMeta,
		writeBundle,
		writeBundleMeta,
		closeBundle,
		closeBundleMeta,
		banner,
		bannerMeta,
		footer,
		footerMeta,
		intro,
		introMeta,
		outro,
		outroMeta,
		watchChange,
		watchChangeMeta,
		closeWatcher,
		closeWatcherMeta,
		hookUsage
	};
	return wrapHandlers(result);
}
function wrapHandlers(plugin) {
	for (const hookName of [
		"buildStart",
		"resolveId",
		"resolveDynamicImport",
		"buildEnd",
		"transform",
		"moduleParsed",
		"load",
		"renderChunk",
		"augmentChunkHash",
		"renderStart",
		"renderError",
		"generateBundle",
		"writeBundle",
		"closeBundle",
		"banner",
		"footer",
		"intro",
		"outro",
		"watchChange",
		"closeWatcher"
	]) {
		const handler = plugin[hookName];
		if (handler) plugin[hookName] = async (...args) => {
			try {
				return await handler(...args);
			} catch (e$5) {
				return error(logPluginError(e$5, plugin.name, {
					hook: hookName,
					id: hookName === "transform" ? args[2] : void 0
				}));
			}
		};
	}
	return plugin;
}
var init_bindingify_plugin = __esm({ "src/plugin/bindingify-plugin.ts"() {
	init_bindingify_build_hooks();
	init_bindingify_output_hooks();
	init_logs();
	init_bindingify_watch_hooks();
	init_hook_usage();
} });

//#endregion
//#region src/plugin/plugin-context-data.ts
var PluginContextData;
var init_plugin_context_data = __esm({ "src/plugin/plugin-context-data.ts"() {
	init_transform_module_info();
	PluginContextData = class {
		moduleOptionMap = new Map();
		resolveOptionsMap = new Map();
		loadModulePromiseMap = new Map();
		loadModulePromiseResolveFnMap = new Map();
		renderedChunkMeta = null;
		updateModuleOption(id$1, option) {
			const existing = this.moduleOptionMap.get(id$1);
			if (existing) {
				if (option.moduleSideEffects != null) existing.moduleSideEffects = option.moduleSideEffects;
				if (option.meta != null) Object.assign(existing.meta, option.meta);
				if (option.invalidate != null) existing.invalidate = option.invalidate;
			} else {
				this.moduleOptionMap.set(id$1, option);
				return option;
			}
			return existing;
		}
		getModuleOption(id$1) {
			const option = this.moduleOptionMap.get(id$1);
			if (!option) {
				const raw = {
					moduleSideEffects: null,
					meta: {}
				};
				this.moduleOptionMap.set(id$1, raw);
				return raw;
			}
			return option;
		}
		getModuleInfo(id$1, context) {
			const bindingInfo = context.getModuleInfo(id$1);
			if (bindingInfo) {
				const info = transformModuleInfo(bindingInfo, this.getModuleOption(id$1));
				return this.proxyModuleInfo(id$1, info);
			}
			return null;
		}
		proxyModuleInfo(id$1, info) {
			let moduleSideEffects = info.moduleSideEffects;
			Object.defineProperty(info, "moduleSideEffects", {
				get() {
					return moduleSideEffects;
				},
				set: (v$1) => {
					this.updateModuleOption(id$1, {
						moduleSideEffects: v$1,
						meta: info.meta,
						invalidate: true
					});
					moduleSideEffects = v$1;
				}
			});
			return info;
		}
		getModuleIds(context) {
			const moduleIds = context.getModuleIds();
			return moduleIds.values();
		}
		saveResolveOptions(options) {
			const index = this.resolveOptionsMap.size;
			this.resolveOptionsMap.set(index, options);
			return index;
		}
		getSavedResolveOptions(receipt) {
			return this.resolveOptionsMap.get(receipt);
		}
		removeSavedResolveOptions(receipt) {
			this.resolveOptionsMap.delete(receipt);
		}
		setRenderChunkMeta(meta) {
			this.renderedChunkMeta = meta;
		}
		getRenderChunkMeta() {
			return this.renderedChunkMeta;
		}
		markModuleLoaded(id$1, _success) {
			const resolve = this.loadModulePromiseResolveFnMap.get(id$1);
			if (resolve) resolve();
		}
		clear() {
			this.renderedChunkMeta = null;
			this.loadModulePromiseMap.clear();
			this.loadModulePromiseResolveFnMap.clear();
		}
	};
} });

//#endregion
//#region src/utils/normalize-string-or-regex.ts
function normalizedStringOrRegex(pattern) {
	if (!pattern) return void 0;
	if (!isReadonlyArray(pattern)) return [pattern];
	return pattern;
}
function isReadonlyArray(input) {
	return Array.isArray(input);
}
var init_normalize_string_or_regex = __esm({ "src/utils/normalize-string-or-regex.ts"() {} });

//#endregion
//#region src/utils/bindingify-input-options.ts
function bindingifyInputOptions(rawPlugins, inputOptions, outputOptions, normalizedOutputPlugins, onLog, logLevel, watchMode) {
	const pluginContextData = new PluginContextData();
	const plugins = rawPlugins.map((plugin) => {
		if ("_parallel" in plugin) return void 0;
		if (plugin instanceof BuiltinPlugin) return bindingifyBuiltInPlugin(plugin);
		return bindingifyPlugin(plugin, inputOptions, outputOptions, pluginContextData, normalizedOutputPlugins, onLog, logLevel, watchMode);
	});
	return {
		input: bindingifyInput(inputOptions.input),
		plugins,
		cwd: inputOptions.cwd ?? process.cwd(),
		external: bindingifyExternal(inputOptions.external),
		resolve: bindingifyResolve(inputOptions.resolve),
		platform: inputOptions.platform,
		shimMissingExports: inputOptions.shimMissingExports,
		logLevel: bindingifyLogLevel(logLevel),
		onLog,
		treeshake: bindingifyTreeshakeOptions(inputOptions.treeshake),
		moduleTypes: inputOptions.moduleTypes,
		define: inputOptions.define ? Object.entries(inputOptions.define) : void 0,
		inject: bindingifyInject(inputOptions.inject),
		experimental: {
			strictExecutionOrder: inputOptions.experimental?.strictExecutionOrder,
			disableLiveBindings: inputOptions.experimental?.disableLiveBindings,
			viteMode: inputOptions.experimental?.viteMode,
			resolveNewUrlToAsset: inputOptions.experimental?.resolveNewUrlToAsset,
			hmr: bindingifyHmr(inputOptions.experimental?.hmr)
		},
		profilerNames: inputOptions?.profilerNames,
		jsx: bindingifyJsx(inputOptions.jsx),
		transform: inputOptions.transform,
		watch: bindingifyWatch(inputOptions.watch),
		dropLabels: inputOptions.dropLabels,
		keepNames: inputOptions.keepNames,
		checks: inputOptions.checks,
		deferSyncScanData: () => {
			let ret = [];
			pluginContextData.moduleOptionMap.forEach((value, key) => {
				if (value.invalidate) ret.push({
					id: key,
					sideEffects: bindingifySideEffects(value.moduleSideEffects)
				});
			});
			return ret;
		},
		makeAbsoluteExternalsRelative: bindingifyMakeAbsoluteExternalsRelative(inputOptions.makeAbsoluteExternalsRelative),
		debug: inputOptions.debug,
		invalidateJsSideCache: pluginContextData.clear.bind(pluginContextData),
		markModuleLoaded: pluginContextData.markModuleLoaded.bind(pluginContextData)
	};
}
function bindingifyHmr(hmr) {
	if (hmr) {
		if (typeof hmr === "boolean") return hmr ? {} : void 0;
		return hmr;
	}
}
function bindingifyExternal(external) {
	if (external) {
		if (typeof external === "function") return (id$1, importer, isResolved) => {
			if (id$1.startsWith("\0")) return false;
			return external(id$1, importer, isResolved) ?? false;
		};
		const externalArr = arraify(external);
		return (id$1, _importer, _isResolved) => {
			return externalArr.some((pat) => {
				if (pat instanceof RegExp) return pat.test(id$1);
				return id$1 === pat;
			});
		};
	}
}
function bindingifyResolve(resolve) {
	if (resolve) {
		const { alias, extensionAlias,...rest } = resolve;
		return {
			alias: alias ? Object.entries(alias).map(([name, replacement]) => ({
				find: name,
				replacements: arraify(replacement)
			})) : void 0,
			extensionAlias: extensionAlias ? Object.entries(extensionAlias).map(([name, value]) => ({
				target: name,
				replacements: value
			})) : void 0,
			...rest
		};
	}
}
function bindingifyInject(inject) {
	if (inject) return Object.entries(inject).map(([alias, item]) => {
		if (Array.isArray(item)) {
			if (item[1] === "*") return {
				tagNamespace: true,
				alias,
				from: item[0]
			};
			return {
				tagNamed: true,
				alias,
				from: item[0],
				imported: item[1]
			};
		} else return {
			tagNamed: true,
			imported: "default",
			alias,
			from: item
		};
	});
}
function bindingifyLogLevel(logLevel) {
	switch (logLevel) {
		case "silent": return import_binding.BindingLogLevel.Silent;
		case "debug": return import_binding.BindingLogLevel.Debug;
		case "warn": return import_binding.BindingLogLevel.Warn;
		case "info": return import_binding.BindingLogLevel.Info;
		default: throw new Error(`Unexpected log level: ${logLevel}`);
	}
}
function bindingifyInput(input) {
	if (input === void 0) return [];
	if (typeof input === "string") return [{ import: input }];
	if (Array.isArray(input)) return input.map((src) => ({ import: src }));
	return Object.entries(input).map(([name, import_path]) => {
		return {
			name,
			import: import_path
		};
	});
}
function bindingifyJsx(input) {
	switch (input) {
		case false: return { type: "Disable" };
		case "react": return { type: "React" };
		case "react-jsx": return { type: "ReactJsx" };
		case "preserve": return { type: "Preserve" };
		case void 0: return void 0;
	}
	if (input.mode === "preserve") return { type: "Preserve" };
	const mode = input.mode ?? "automatic";
	return {
		type: "Enable",
		field0: {
			runtime: mode,
			importSource: mode === "classic" ? input.importSource : mode === "automatic" ? input.jsxImportSource : void 0,
			pragma: input.factory,
			pragmaFrag: input.fragment,
			development: input.development,
			refresh: input.refresh
		}
	};
}
function bindingifyWatch(watch$1) {
	if (watch$1) return {
		buildDelay: watch$1.buildDelay,
		skipWrite: watch$1.skipWrite,
		include: normalizedStringOrRegex(watch$1.include),
		exclude: normalizedStringOrRegex(watch$1.exclude)
	};
}
function bindingifyTreeshakeOptions(config) {
	if (config === false) return void 0;
	if (config === true || config === void 0) return { moduleSideEffects: true };
	let normalizedConfig = {
		moduleSideEffects: true,
		annotations: config.annotations,
		manualPureFunctions: config.manualPureFunctions,
		unknownGlobalSideEffects: config.unknownGlobalSideEffects
	};
	if (config.moduleSideEffects === void 0) normalizedConfig.moduleSideEffects = true;
	else if (config.moduleSideEffects === "no-external") normalizedConfig.moduleSideEffects = [{
		external: true,
		sideEffects: false
	}, {
		external: false,
		sideEffects: true
	}];
	else normalizedConfig.moduleSideEffects = config.moduleSideEffects;
	return normalizedConfig;
}
function bindingifyMakeAbsoluteExternalsRelative(makeAbsoluteExternalsRelative) {
	if (makeAbsoluteExternalsRelative === "ifRelativeSource") return { type: "IfRelativeSource" };
	if (typeof makeAbsoluteExternalsRelative === "boolean") return {
		type: "Bool",
		field0: makeAbsoluteExternalsRelative
	};
}
var init_bindingify_input_options = __esm({ "src/utils/bindingify-input-options.ts"() {
	init_constructors();
	init_utils();
	init_bindingify_plugin();
	init_plugin_context_data();
	init_misc();
	init_normalize_string_or_regex();
	init_transform_side_effects();
} });

//#endregion
//#region src/utils/plugin/index.ts
var isPluginHookName;
var init_plugin$1 = __esm({ "src/utils/plugin/index.ts"() {
	init_plugin$2();
	isPluginHookName = function() {
		const PLUGIN_HOOK_NAMES_SET = new Set(ENUMERATED_PLUGIN_HOOK_NAMES);
		return function isPluginHookName$1(hookName) {
			return PLUGIN_HOOK_NAMES_SET.has(hookName);
		};
	}();
} });

//#endregion
//#region src/utils/compose-js-plugins.ts
function isUnsupportedHooks(hookName) {
	return unsupportedHooks.has(hookName);
}
function createComposedPlugin(plugins) {
	const names = [];
	const batchedHooks = {};
	plugins.forEach((plugin, index) => {
		const pluginName = plugin.name || `Anonymous(index: ${index})`;
		names.push(pluginName);
		t$5(plugin).forEach((pluginProp) => {
			if (isUnsupportedHooks(pluginProp)) throw new Error(`Failed to compose js plugins. Plugin ${pluginName} has an unsupported hook: ${pluginProp}`);
			if (!isPluginHookName(pluginProp)) return;
			switch (pluginProp) {
				case "buildStart": {
					const handlers = batchedHooks.buildStart ?? [];
					batchedHooks.buildStart = handlers;
					if (plugin.buildStart) handlers.push([plugin.buildStart, plugin]);
					break;
				}
				case "load": {
					const handlers = batchedHooks.load ?? [];
					batchedHooks.load = handlers;
					if (plugin.load) handlers.push([plugin.load, plugin]);
					break;
				}
				case "transform": {
					const handlers = batchedHooks.transform ?? [];
					batchedHooks.transform = handlers;
					if (plugin.transform) handlers.push([plugin.transform, plugin]);
					break;
				}
				case "resolveId": {
					const handlers = batchedHooks.resolveId ?? [];
					batchedHooks.resolveId = handlers;
					if (plugin.resolveId) handlers.push([plugin.resolveId, plugin]);
					break;
				}
				case "buildEnd": {
					const handlers = batchedHooks.buildEnd ?? [];
					batchedHooks.buildEnd = handlers;
					if (plugin.buildEnd) handlers.push([plugin.buildEnd, plugin]);
					break;
				}
				case "renderChunk": {
					const handlers = batchedHooks.renderChunk ?? [];
					batchedHooks.renderChunk = handlers;
					if (plugin.renderChunk) handlers.push([plugin.renderChunk, plugin]);
					break;
				}
				case "banner":
				case "footer":
				case "intro":
				case "outro": {
					const hook = plugin[pluginProp];
					if (hook) (batchedHooks[pluginProp] ??= []).push([hook, plugin]);
					break;
				}
				case "closeBundle": {
					const handlers = batchedHooks.closeBundle ?? [];
					batchedHooks.closeBundle = handlers;
					if (plugin.closeBundle) handlers.push([plugin.closeBundle, plugin]);
					break;
				}
				case "watchChange": {
					const handlers = batchedHooks.watchChange ?? [];
					batchedHooks.watchChange = handlers;
					if (plugin.watchChange) handlers.push([plugin.watchChange, plugin]);
					break;
				}
				case "closeWatcher": {
					const handlers = batchedHooks.closeWatcher ?? [];
					batchedHooks.closeWatcher = handlers;
					if (plugin.closeWatcher) handlers.push([plugin.closeWatcher, plugin]);
					break;
				}
				default: {}
			}
		});
	});
	const composed = { name: `Composed(${names.join(", ")})` };
	const createFixedPluginResolveFnMap = new Map();
	function applyFixedPluginResolveFn(ctx, plugin) {
		const createFixedPluginResolveFn = createFixedPluginResolveFnMap.get(plugin);
		if (createFixedPluginResolveFn) ctx.resolve = createFixedPluginResolveFn(ctx, ctx.resolve.bind(ctx));
		return ctx;
	}
	if (batchedHooks.resolveId) {
		const batchedHandlers = batchedHooks.resolveId;
		const handlerSymbols = batchedHandlers.map(([_handler, plugin]) => Symbol(plugin.name ?? `Anonymous`));
		for (let handlerIdx = 0; handlerIdx < batchedHandlers.length; handlerIdx++) {
			const [_handler, plugin] = batchedHandlers[handlerIdx];
			const handlerSymbol = handlerSymbols[handlerIdx];
			const createFixedPluginResolveFn = (ctx, resolve) => {
				return (source, importer, rawContextResolveOptions) => {
					const contextResolveOptions = rawContextResolveOptions ?? {};
					if (contextResolveOptions.skipSelf) {
						contextResolveOptions[SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF] = handlerSymbol;
						contextResolveOptions.skipSelf = false;
					}
					return resolve(source, importer, contextResolveOptions);
				};
			};
			createFixedPluginResolveFnMap.set(plugin, createFixedPluginResolveFn);
		}
		composed.resolveId = async function(source, importer, rawHookResolveIdOptions) {
			const hookResolveIdOptions = rawHookResolveIdOptions;
			const symbolForCallerThatSkipSelf = hookResolveIdOptions?.[SYMBOL_FOR_RESOLVE_CALLER_THAT_SKIP_SELF];
			for (let handlerIdx = 0; handlerIdx < batchedHandlers.length; handlerIdx++) {
				const [handler, plugin] = batchedHandlers[handlerIdx];
				const handlerSymbol = handlerSymbols[handlerIdx];
				if (symbolForCallerThatSkipSelf === handlerSymbol) continue;
				const { handler: handlerFn } = normalizeHook(handler);
				const result = await handlerFn.call(applyFixedPluginResolveFn(this, plugin), source, importer, rawHookResolveIdOptions);
				if (!isNullish(result)) return result;
			}
		};
	}
	t$5(batchedHooks).forEach((hookName) => {
		switch (hookName) {
			case "resolveId": break;
			case "buildStart": {
				if (batchedHooks.buildStart) {
					const batchedHandlers = batchedHooks.buildStart;
					composed.buildStart = async function(options) {
						await Promise.all(batchedHandlers.map(([handler, plugin]) => {
							const { handler: handlerFn } = normalizeHook(handler);
							return handlerFn.call(applyFixedPluginResolveFn(this, plugin), options);
						}));
					};
				}
				break;
			}
			case "load": {
				if (batchedHooks.load) {
					const batchedHandlers = batchedHooks.load;
					composed.load = async function(id$1) {
						for (const [handler, plugin] of batchedHandlers) {
							const { handler: handlerFn } = normalizeHook(handler);
							const result = await handlerFn.call(applyFixedPluginResolveFn(this, plugin), id$1);
							if (!isNullish(result)) return result;
						}
					};
				}
				break;
			}
			case "transform": {
				if (batchedHooks.transform) {
					const batchedHandlers = batchedHooks.transform;
					composed.transform = async function(initialCode, id$1, moduleType$1) {
						let code$1 = initialCode;
						let moduleSideEffects = void 0;
						function updateOutput(newCode, newModuleSideEffects) {
							code$1 = newCode;
							moduleSideEffects = newModuleSideEffects ?? void 0;
						}
						for (const [handler, plugin] of batchedHandlers) {
							const { handler: handlerFn } = normalizeHook(handler);
							this.getCombinedSourcemap = () => {
								throw new Error(`The getCombinedSourcemap is not implement in transform hook at composedJsPlugins`);
							};
							const result = await handlerFn.call(applyFixedPluginResolveFn(this, plugin), code$1, id$1, moduleType$1);
							if (!isNullish(result)) {
								if (typeof result === "string") updateOutput(result);
								else if (result.code) updateOutput(result.code, result.moduleSideEffects);
							}
						}
						return {
							code: code$1,
							moduleSideEffects
						};
					};
				}
				break;
			}
			case "buildEnd": {
				if (batchedHooks.buildEnd) {
					const batchedHandlers = batchedHooks.buildEnd;
					composed.buildEnd = async function(err) {
						await Promise.all(batchedHandlers.map(([handler, plugin]) => {
							const { handler: handlerFn } = normalizeHook(handler);
							return handlerFn.call(applyFixedPluginResolveFn(this, plugin), err);
						}));
					};
				}
				break;
			}
			case "renderChunk": {
				if (batchedHooks.renderChunk) {
					const batchedHandlers = batchedHooks.renderChunk;
					composed.renderChunk = async function(code$1, chunk, options, meta) {
						for (const [handler, plugin] of batchedHandlers) {
							const { handler: handlerFn } = normalizeHook(handler);
							const result = await handlerFn.call(applyFixedPluginResolveFn(this, plugin), code$1, chunk, options, meta);
							if (!isNullish(result)) return result;
						}
					};
				}
				break;
			}
			case "banner":
			case "footer":
			case "intro":
			case "outro": {
				const hooks = batchedHooks[hookName];
				if (hooks?.length) composed[hookName] = async function(chunk) {
					const ret = [];
					for (const [hook, plugin] of hooks) {
						const { handler } = normalizeHook(hook);
						ret.push(typeof handler === "string" ? handler : await handler.call(applyFixedPluginResolveFn(this, plugin), chunk));
					}
					return ret.join("\n");
				};
				break;
			}
			case "closeBundle": {
				if (batchedHooks.closeBundle) {
					const batchedHandlers = batchedHooks.closeBundle;
					composed.closeBundle = async function() {
						await Promise.all(batchedHandlers.map(([handler, plugin]) => {
							const { handler: handlerFn } = normalizeHook(handler);
							return handlerFn.call(applyFixedPluginResolveFn(this, plugin));
						}));
					};
				}
				break;
			}
			case "watchChange": {
				if (batchedHooks.watchChange) {
					const batchedHandlers = batchedHooks.watchChange;
					composed.watchChange = async function(id$1, event) {
						await Promise.all(batchedHandlers.map(([handler, plugin]) => {
							const { handler: handlerFn } = normalizeHook(handler);
							return handlerFn.call(applyFixedPluginResolveFn(this, plugin), id$1, event);
						}));
					};
				}
				break;
			}
			case "closeWatcher": {
				if (batchedHooks.closeWatcher) {
					const batchedHandlers = batchedHooks.closeWatcher;
					composed.closeWatcher = async function() {
						await Promise.all(batchedHandlers.map(([handler, plugin]) => {
							const { handler: handlerFn } = normalizeHook(handler);
							return handlerFn.call(applyFixedPluginResolveFn(this, plugin));
						}));
					};
				}
				break;
			}
			default: {}
		}
	});
	return composed;
}
function isComposablePlugin(plugin) {
	if (plugin instanceof BuiltinPlugin) return false;
	if ("_parallel" in plugin) return false;
	const hasNotComposablePattern = t$5(plugin).some((hookName) => {
		if (!isPluginHookName(hookName)) return false;
		const OK_TO_COMPOSE = false;
		if (isUnsupportedHooks(hookName)) return !OK_TO_COMPOSE;
		if (plugin[hookName]) {
			const { meta } = normalizeHook(plugin[hookName]);
			if (meta.order === "pre" || meta.order === "post") return !OK_TO_COMPOSE;
		}
		return OK_TO_COMPOSE;
	});
	if (hasNotComposablePattern) return false;
	return true;
}
function composeJsPlugins(plugins) {
	const newPlugins = [];
	const toBeComposed = [];
	plugins.forEach((plugin) => {
		if (isComposablePlugin(plugin)) toBeComposed.push(plugin);
		else {
			if (toBeComposed.length > 0) {
				if (toBeComposed.length > 1) newPlugins.push(createComposedPlugin(toBeComposed));
				else newPlugins.push(toBeComposed[0]);
				toBeComposed.length = 0;
			}
			newPlugins.push(plugin);
		}
	});
	if (toBeComposed.length > 0) {
		if (toBeComposed.length > 1) newPlugins.push(createComposedPlugin(toBeComposed));
		else newPlugins.push(toBeComposed[0]);
		toBeComposed.length = 0;
	}
	return newPlugins;
}
var unsupportedHookName, unsupportedHooks;
var init_compose_js_plugins = __esm({ "src/utils/compose-js-plugins.ts"() {
	init_dist();
	init_constructors();
	init_plugin_context$1();
	init_misc();
	init_normalize_hook();
	init_plugin$1();
	unsupportedHookName = [
		"augmentChunkHash",
		"generateBundle",
		"moduleParsed",
		"onLog",
		"options",
		"outputOptions",
		"renderError",
		"renderStart",
		"resolveDynamicImport",
		"writeBundle"
	];
	unsupportedHooks = new Set(unsupportedHookName);
} });

//#endregion
//#region src/utils/initialize-parallel-plugins.ts
async function initializeParallelPlugins(plugins) {
	const pluginInfos = [];
	for (const [index, plugin] of plugins.entries()) if ("_parallel" in plugin) {
		const { fileUrl, options } = plugin._parallel;
		pluginInfos.push({
			index,
			fileUrl,
			options
		});
	}
	if (pluginInfos.length <= 0) return void 0;
	const count = Math.min(availableParallelism(), 8);
	const parallelJsPluginRegistry = new import_binding.ParallelJsPluginRegistry(count);
	const registryId = parallelJsPluginRegistry.id;
	const workers = await initializeWorkers(registryId, count, pluginInfos);
	const stopWorkers = async () => {
		await Promise.all(workers.map((worker) => worker.terminate()));
	};
	return {
		registry: parallelJsPluginRegistry,
		stopWorkers
	};
}
function initializeWorkers(registryId, count, pluginInfos) {
	return Promise.all(Array.from({ length: count }, (_, i$22) => initializeWorker(registryId, pluginInfos, i$22)));
}
async function initializeWorker(registryId, pluginInfos, threadNumber) {
	const urlString = import.meta.resolve("#parallel-plugin-worker");
	const workerData$1 = {
		registryId,
		pluginInfos,
		threadNumber
	};
	let worker;
	try {
		worker = new Worker(new URL(urlString), { workerData: workerData$1 });
		worker.unref();
		await new Promise((resolve, reject) => {
			worker.once("message", async (message) => {
				if (message.type === "error") reject(message.error);
				else resolve();
			});
		});
		return worker;
	} catch (e$5) {
		worker?.terminate();
		throw e$5;
	}
}
var init_initialize_parallel_plugins = __esm({ "src/utils/initialize-parallel-plugins.ts"() {} });

//#endregion
//#region src/utils/create-bundler-option.ts
async function createBundlerOptions(inputOptions, outputOptions, watchMode, isClose) {
	const inputPlugins = await normalizePluginOption(inputOptions.plugins);
	const outputPlugins = await normalizePluginOption(outputOptions.plugins);
	const logLevel = inputOptions.logLevel || LOG_LEVEL_INFO;
	const onLog = getLogger(getObjectPlugins(inputPlugins), getOnLog(inputOptions, logLevel), logLevel, watchMode);
	if (!isClose) outputOptions = PluginDriver.callOutputOptionsHook([...inputPlugins, ...outputPlugins], outputOptions, onLog, logLevel, watchMode);
	const normalizedOutputPlugins = await normalizePluginOption(outputOptions.plugins);
	let plugins = [...normalizePlugins(inputPlugins, ANONYMOUS_PLUGIN_PREFIX), ...checkOutputPluginOption(normalizePlugins(normalizedOutputPlugins, ANONYMOUS_OUTPUT_PLUGIN_PREFIX), onLog)];
	if (inputOptions.experimental?.enableComposingJsPlugins ?? false) plugins = composeJsPlugins(plugins);
	const parallelPluginInitResult = await initializeParallelPlugins(plugins);
	try {
		const bindingInputOptions = bindingifyInputOptions(plugins, inputOptions, outputOptions, normalizedOutputPlugins, onLog, logLevel, watchMode);
		const bindingOutputOptions = bindingifyOutputOptions(outputOptions);
		return {
			bundlerOptions: {
				inputOptions: bindingInputOptions,
				outputOptions: bindingOutputOptions,
				parallelPluginsRegistry: parallelPluginInitResult?.registry
			},
			inputOptions,
			onLog,
			stopWorkers: parallelPluginInitResult?.stopWorkers
		};
	} catch (e$5) {
		await parallelPluginInitResult?.stopWorkers();
		throw e$5;
	}
}
var init_create_bundler_option = __esm({ "src/utils/create-bundler-option.ts"() {
	init_logger();
	init_logging();
	init_plugin_driver();
	init_bindingify_input_options();
	init_bindingify_output_options();
	init_compose_js_plugins();
	init_initialize_parallel_plugins();
	init_normalize_plugin_option();
} });

//#endregion
//#region src/utils/create-bundler.ts
async function createBundler(inputOptions, outputOptions, isClose) {
	const option = await createBundlerOptions(inputOptions, outputOptions, false, isClose);
	if (asyncRuntimeShutdown) (0, import_binding.startAsyncRuntime)();
	try {
		return {
			bundler: new import_binding.Bundler(option.bundlerOptions),
			stopWorkers: option.stopWorkers,
			shutdown: () => {
				(0, import_binding.shutdownAsyncRuntime)();
				asyncRuntimeShutdown = true;
			}
		};
	} catch (e$5) {
		await option.stopWorkers?.();
		throw e$5;
	}
}
var asyncRuntimeShutdown;
var init_create_bundler = __esm({ "src/utils/create-bundler.ts"() {
	init_create_bundler_option();
	asyncRuntimeShutdown = false;
} });

//#endregion
//#region src/api/rolldown/rolldown-build.ts
var RolldownBuild;
var init_rolldown_build = __esm({ "src/api/rolldown/rolldown-build.ts"() {
	init_create_bundler();
	init_transform_to_rollup_output();
	init_validator();
	Symbol.asyncDispose ??= Symbol("Symbol.asyncDispose");
	RolldownBuild = class {
		#inputOptions;
		#bundler;
		constructor(inputOptions) {
			this.#inputOptions = inputOptions;
		}
		get closed() {
			return this.#bundler?.bundler.closed ?? false;
		}
		async #getBundlerWithStopWorker(outputOptions, isClose) {
			if (this.#bundler) await this.#bundler.stopWorkers?.();
			return this.#bundler = await createBundler(this.#inputOptions, outputOptions, isClose);
		}
		async generate(outputOptions = {}) {
			validateOption("output", outputOptions);
			const { bundler } = await this.#getBundlerWithStopWorker(outputOptions);
			const output = await bundler.generate();
			return transformToRollupOutput(output);
		}
		async write(outputOptions = {}) {
			validateOption("output", outputOptions);
			const { bundler } = await this.#getBundlerWithStopWorker(outputOptions);
			const output = await bundler.write();
			return transformToRollupOutput(output);
		}
		async close() {
			const { bundler, stopWorkers, shutdown } = await this.#getBundlerWithStopWorker({}, true);
			await stopWorkers?.();
			await bundler.close();
			shutdown();
		}
		async [Symbol.asyncDispose]() {
			await this.close();
		}
		async generateHmrPatch(changedFiles) {
			return this.#bundler?.bundler.generateHmrPatch(changedFiles);
		}
		get watchFiles() {
			return this.#bundler?.bundler.watchFiles ?? [];
		}
	};
} });

//#endregion
//#region src/api/rolldown/index.ts
var rolldown;
var init_rolldown = __esm({ "src/api/rolldown/index.ts"() {
	init_plugin_driver();
	init_validator();
	init_rolldown_build();
	rolldown = async (input) => {
		validateOption("input", input);
		const inputOptions = await PluginDriver.callOptionsHook(input);
		return new RolldownBuild(inputOptions);
	};
} });

//#endregion
//#region src/api/build.ts
async function build(options) {
	if (Array.isArray(options)) return Promise.all(options.map((opts) => build(opts)));
	else {
		const { output, write = true,...inputOptions } = options;
		const build$1 = await rolldown(inputOptions);
		try {
			if (write) return await build$1.write(output);
			else return await build$1.generate(output);
		} finally {
			await build$1.close();
		}
	}
}
var init_build = __esm({ "src/api/build.ts"() {
	init_rolldown();
} });

//#endregion
//#region src/api/watch/watch-emitter.ts
var WatcherEmitter;
var init_watch_emitter = __esm({ "src/api/watch/watch-emitter.ts"() {
	init_error();
	WatcherEmitter = class {
		listeners = new Map();
		timer;
		constructor() {
			this.timer = setInterval(
				() => {},
				1e9
				/* Low power usage */
);
		}
		on(event, listener) {
			const listeners = this.listeners.get(event);
			if (listeners) listeners.push(listener);
			else this.listeners.set(event, [listener]);
			return this;
		}
		async onEvent(event) {
			const listeners = this.listeners.get(event.eventKind());
			if (listeners) switch (event.eventKind()) {
				case "close":
				case "restart":
					for (const listener of listeners) await listener();
					break;
				case "event":
					for (const listener of listeners) {
						const code$1 = event.bundleEventKind();
						switch (code$1) {
							case "BUNDLE_END":
								const { duration, output } = event.bundleEndData();
								await listener({
									code: "BUNDLE_END",
									duration,
									output: [output]
								});
								break;
							case "ERROR":
								const errors = event.errors();
								await listener({
									code: "ERROR",
									error: normalizeErrors(errors)
								});
								break;
							default:
								await listener({ code: code$1 });
								break;
						}
					}
					break;
				case "change":
					for (const listener of listeners) {
						const { path: path$1, kind } = event.watchChangeData();
						await listener(path$1, { event: kind });
					}
					break;
				default: throw new Error(`Unknown event: ${event}`);
			}
		}
		async close() {
			clearInterval(this.timer);
		}
	};
} });

//#endregion
//#region src/api/watch/watcher.ts
async function createWatcher(emitter, input) {
	const options = arraify(input);
	const bundlerOptions = await Promise.all(options.map((option) => arraify(option.output || {}).map(async (output) => {
		const inputOptions = await PluginDriver.callOptionsHook(option, true);
		return createBundlerOptions(inputOptions, output, true);
	})).flat());
	const notifyOptions = getValidNotifyOption(bundlerOptions);
	const bindingWatcher = new import_binding.BindingWatcher(bundlerOptions.map((option) => option.bundlerOptions), notifyOptions);
	const watcher = new Watcher(emitter, bindingWatcher, bundlerOptions.map((option) => option.stopWorkers));
	watcher.start();
}
function getValidNotifyOption(bundlerOptions) {
	let result;
	for (const option of bundlerOptions) if (option.inputOptions.watch) {
		const notifyOption = option.inputOptions.watch.notify;
		if (notifyOption) if (result) {
			option.onLog(LOG_LEVEL_WARN, logMultiplyNotifyOption());
			return result;
		} else result = notifyOption;
	}
}
var Watcher;
var init_watcher = __esm({ "src/api/watch/watcher.ts"() {
	init_logging();
	init_logs();
	init_plugin_driver();
	init_create_bundler_option();
	init_misc();
	Watcher = class {
		closed;
		inner;
		emitter;
		stopWorkers;
		constructor(emitter, inner, stopWorkers) {
			this.closed = false;
			this.inner = inner;
			this.emitter = emitter;
			const originClose = emitter.close.bind(emitter);
			emitter.close = async () => {
				await this.close();
				originClose();
			};
			this.stopWorkers = stopWorkers;
		}
		async close() {
			if (this.closed) return;
			this.closed = true;
			for (const stop of this.stopWorkers) await stop?.();
			await this.inner.close();
			(0, import_binding.shutdownAsyncRuntime)();
		}
		start() {
			process.nextTick(() => this.inner.start(this.emitter.onEvent.bind(this.emitter)));
		}
	};
} });

//#endregion
//#region src/api/watch/index.ts
var watch;
var init_watch = __esm({ "src/api/watch/index.ts"() {
	init_watch_emitter();
	init_watcher();
	watch = (input) => {
		const emitter = new WatcherEmitter();
		createWatcher(emitter, input);
		return emitter;
	};
} });

//#endregion
//#region src/plugin/with-filter.ts
function withFilterImpl(pluginOption, filterObjectList) {
	if (isPromiseLike(pluginOption)) return pluginOption.then((p$3) => withFilter(p$3, filterObjectList));
	if (pluginOption == false || pluginOption == null) return pluginOption;
	if (Array.isArray(pluginOption)) return pluginOption.map((p$3) => withFilter(p$3, filterObjectList));
	let plugin = pluginOption;
	let filterObjectIndex = findMatchedFilterObject(plugin.name, filterObjectList);
	if (filterObjectIndex === -1) return plugin;
	let filterObject = filterObjectList[filterObjectIndex];
	Object.keys(plugin).forEach((key) => {
		switch (key) {
			case "transform":
			case "resolveId":
			case "load":
				if (!plugin[key]) return;
				if (typeof plugin[key] === "object") plugin[key].filter = filterObject[key] ?? plugin[key].filter;
				else plugin[key] = {
					handler: plugin[key],
					filter: filterObject[key]
				};
				break;
			default: break;
		}
	});
	return plugin;
}
function withFilter(pluginOption, filterObject) {
	return withFilterImpl(pluginOption, arraify(filterObject));
}
function findMatchedFilterObject(pluginName, overrideFilterObjectList) {
	if (overrideFilterObjectList.length === 1 && overrideFilterObjectList[0].pluginNamePattern === void 0) return 0;
	for (let i$22 = 0; i$22 < overrideFilterObjectList.length; i$22++) for (let j$1 = 0; j$1 < (overrideFilterObjectList[i$22].pluginNamePattern ?? []).length; j$1++) {
		let pattern = overrideFilterObjectList[i$22].pluginNamePattern[j$1];
		if (typeof pattern === "string" && pattern === pluginName) return i$22;
		else if (pattern instanceof RegExp && pattern.test(pluginName)) return i$22;
	}
	return -1;
}
var init_with_filter = __esm({ "src/plugin/with-filter.ts"() {
	init_misc();
} });

//#endregion
//#region src/plugin/index.ts
var init_plugin = __esm({ "src/plugin/index.ts"() {
	init_with_filter();
} });

//#endregion
//#region src/utils/define-config.ts
function defineConfig(config) {
	return config;
}
var init_define_config = __esm({ "src/utils/define-config.ts"() {} });

//#endregion
//#region src/index.ts
var VERSION;
var init_src = __esm({ "src/index.ts"() {
	init_build();
	init_rolldown();
	init_watch();
	init_plugin();
	init_define_config();
	VERSION = version;
} });

//#endregion
export { BuiltinPlugin, PluginContextData, VERSION, arraify, bindingifyPlugin, build, buildImportAnalysisPlugin, composeJsPlugins, createBundler, defineConfig, description, dynamicImportVarsPlugin, getInputCliKeys, getJsonSchema, getOutputCliKeys, handleOutputErrors, importGlobPlugin, init_bindingify_plugin, init_compose_js_plugins, init_constructors, init_create_bundler, init_misc, init_normalize_string_or_regex, init_plugin_context_data, init_rolldown, init_src, init_transform_to_rollup_output, init_validator, init_watch, isolatedDeclarationPlugin, jsonPlugin, loadFallbackPlugin, manifestPlugin, moduleFederationPlugin, modulePreloadPolyfillPlugin, normalizedStringOrRegex, reportPlugin, rolldown, validateCliOptions, version, viteResolvePlugin, wasmFallbackPlugin, wasmHelperPlugin, watch, withFilter };
