// unplugin-lightningcss@0.3.3/dist/transform-Fd81TQ13.js
import { Buffer } from "node:buffer";
import { readFile } from "node:fs/promises";

//#region src/core/transform.ts
const postfixRE = /[#?].*$/s;
function cleanUrl(url) {
	return url.replace(postfixRE, "");
}
async function transformCss(id, code, options) {
	const filename = cleanUrl(id);
	const { transform } = await import("lightningcss");
	const res = transform({
		...options,
		filename,
		code: Buffer.from(code)
	});
	return {
		code: res.code.toString(),
		map: "map" in res ? res.map?.toString() : void 0
	};
}
async function transformCssModule(id, options) {
	const actualId = id.replace(/\?css_module$/, "");
	const code = await readFile(actualId, "utf-8");
	const filename = cleanUrl(actualId);
	const { transform } = await import("lightningcss");
	const res = transform({
		cssModules: true,
		...options,
		filename,
		code: Buffer.from(code)
	});
	const compiledId = actualId.replaceAll("\\", "/").replace(/\.module\.css$/, ".module_built.css");
	return {
		code: res.code.toString(),
		map: "map" in res ? res.map?.toString() : void 0,
		id: compiledId,
		exports: res.exports ? Object.entries(res.exports).map(([name, { name: className }]) => `export const ${name} = "${className}";`).join("\n") : ""
	};
}

//#endregion
export { transformCss, transformCssModule };
